{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intro to Pandas and Matplotlib Part I"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quiz 4 Answers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['abc', 'def', 'ghi', 'jkl']\n"
     ]
    }
   ],
   "source": [
    "x = 'abc def   ghi\\njkl\\n'\n",
    "print(x.strip().split())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    }
   ],
   "source": [
    "def compute(n):\n",
    "    a, b = 0, 1\n",
    "    for _ in range(n):\n",
    "        a, b = b, a + b\n",
    "    return b-a\n",
    "\n",
    "\n",
    "print(compute(6))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "num_list.txt\n",
    "\n",
    "5,7,3\n",
    "\n",
    "2,4,2\n",
    "\n",
    "4,8,6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "frh = open('num_list.txt', 'r')\n",
    "fwh = open('result.txt', 'w')\n",
    "result = 0\n",
    "for line in frh:\n",
    "    n = line.strip().split(',')\n",
    "    result += (int(n[0]) + int(n[1])) / int(n[2])\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Answer: 9.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating objects in Pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas is  a great library for manipulating data that we are used to dealing with in biology a.k.a the excel spreadsheet.  Pandas can also create basic plots utilizing the power of the matplotlib API which we will also cover in more detail."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For more info (and to see where I'm pulling material for this tutorial) check out [10 min to Pandas](http://pandas.pydata.org/pandas-docs/stable/10min.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "% matplotlib inline\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are two main object types in Pandas, 1) the series and 2) the dataframe.  A series looks a lot like a list and can accept mixed data types.  It can also be indexed like a list.  And like a list, series objects are mutable and therefore item assingment can take place."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    1\n",
      "1    3\n",
      "2    5\n",
      "3    a\n",
      "4    6\n",
      "5    8\n",
      "dtype: object\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "s = pd.Series([1,3,5,'a',6,8])\n",
    "print(s)\n",
    "#indexing looks is similar to a list\n",
    "print(s[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    2\n",
      "1    3\n",
      "2    5\n",
      "3    a\n",
      "4    6\n",
      "5    8\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "#item reassignment\n",
    "s[0] = 2\n",
    "print(s)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get a list of series methods, you can use tab completetion in the notebook or dir().  Try this below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['T',\n",
       " '_AXIS_ALIASES',\n",
       " '_AXIS_IALIASES',\n",
       " '_AXIS_LEN',\n",
       " '_AXIS_NAMES',\n",
       " '_AXIS_NUMBERS',\n",
       " '_AXIS_ORDERS',\n",
       " '_AXIS_REVERSED',\n",
       " '_AXIS_SLICEMAP',\n",
       " '__abs__',\n",
       " '__add__',\n",
       " '__and__',\n",
       " '__array__',\n",
       " '__array_prepare__',\n",
       " '__array_priority__',\n",
       " '__array_wrap__',\n",
       " '__bool__',\n",
       " '__bytes__',\n",
       " '__class__',\n",
       " '__contains__',\n",
       " '__copy__',\n",
       " '__deepcopy__',\n",
       " '__delattr__',\n",
       " '__delitem__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__div__',\n",
       " '__divmod__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__finalize__',\n",
       " '__float__',\n",
       " '__floordiv__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattr__',\n",
       " '__getattribute__',\n",
       " '__getitem__',\n",
       " '__getstate__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__iadd__',\n",
       " '__iand__',\n",
       " '__ifloordiv__',\n",
       " '__imod__',\n",
       " '__imul__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__int__',\n",
       " '__invert__',\n",
       " '__ior__',\n",
       " '__ipow__',\n",
       " '__isub__',\n",
       " '__iter__',\n",
       " '__itruediv__',\n",
       " '__ixor__',\n",
       " '__le__',\n",
       " '__len__',\n",
       " '__long__',\n",
       " '__lt__',\n",
       " '__mod__',\n",
       " '__module__',\n",
       " '__mul__',\n",
       " '__ne__',\n",
       " '__neg__',\n",
       " '__new__',\n",
       " '__nonzero__',\n",
       " '__or__',\n",
       " '__pow__',\n",
       " '__radd__',\n",
       " '__rand__',\n",
       " '__rdiv__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__rfloordiv__',\n",
       " '__rmod__',\n",
       " '__rmul__',\n",
       " '__ror__',\n",
       " '__round__',\n",
       " '__rpow__',\n",
       " '__rsub__',\n",
       " '__rtruediv__',\n",
       " '__rxor__',\n",
       " '__setattr__',\n",
       " '__setitem__',\n",
       " '__setstate__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__sub__',\n",
       " '__subclasshook__',\n",
       " '__truediv__',\n",
       " '__unicode__',\n",
       " '__weakref__',\n",
       " '__xor__',\n",
       " '_accessors',\n",
       " '_add_numeric_operations',\n",
       " '_add_series_only_operations',\n",
       " '_add_series_or_dataframe_operations',\n",
       " '_agg_by_level',\n",
       " '_agg_doc',\n",
       " '_aggregate',\n",
       " '_aggregate_multiple_funcs',\n",
       " '_align_frame',\n",
       " '_align_series',\n",
       " '_allow_index_ops',\n",
       " '_at',\n",
       " '_binop',\n",
       " '_box_item_values',\n",
       " '_builtin_table',\n",
       " '_can_hold_na',\n",
       " '_check_inplace_setting',\n",
       " '_check_is_chained_assignment_possible',\n",
       " '_check_percentile',\n",
       " '_check_setitem_copy',\n",
       " '_clear_item_cache',\n",
       " '_clip_with_one_bound',\n",
       " '_clip_with_scalar',\n",
       " '_consolidate',\n",
       " '_consolidate_inplace',\n",
       " '_construct_axes_dict',\n",
       " '_construct_axes_dict_for_slice',\n",
       " '_construct_axes_dict_from',\n",
       " '_construct_axes_from_arguments',\n",
       " '_constructor',\n",
       " '_constructor_expanddim',\n",
       " '_constructor_sliced',\n",
       " '_convert',\n",
       " '_create_indexer',\n",
       " '_cython_table',\n",
       " '_deprecations',\n",
       " '_dir_additions',\n",
       " '_dir_deletions',\n",
       " '_drop_axis',\n",
       " '_expand_axes',\n",
       " '_formatting_values',\n",
       " '_from_axes',\n",
       " '_get_axis',\n",
       " '_get_axis_name',\n",
       " '_get_axis_number',\n",
       " '_get_axis_resolvers',\n",
       " '_get_block_manager_axis',\n",
       " '_get_bool_data',\n",
       " '_get_cacher',\n",
       " '_get_index_resolvers',\n",
       " '_get_item_cache',\n",
       " '_get_numeric_data',\n",
       " '_get_value',\n",
       " '_get_values',\n",
       " '_get_values_tuple',\n",
       " '_get_with',\n",
       " '_gotitem',\n",
       " '_iat',\n",
       " '_iget_item_cache',\n",
       " '_iloc',\n",
       " '_index',\n",
       " '_indexed_same',\n",
       " '_info_axis',\n",
       " '_info_axis_name',\n",
       " '_info_axis_number',\n",
       " '_init_mgr',\n",
       " '_internal_names',\n",
       " '_internal_names_set',\n",
       " '_is_builtin_func',\n",
       " '_is_cached',\n",
       " '_is_cython_func',\n",
       " '_is_datelike_mixed_type',\n",
       " '_is_mixed_type',\n",
       " '_is_numeric_mixed_type',\n",
       " '_is_view',\n",
       " '_ix',\n",
       " '_ixs',\n",
       " '_loc',\n",
       " '_maybe_cache_changed',\n",
       " '_maybe_update_cacher',\n",
       " '_metadata',\n",
       " '_needs_reindex_multi',\n",
       " '_obj_with_exclusions',\n",
       " '_protect_consolidate',\n",
       " '_reduce',\n",
       " '_reindex_axes',\n",
       " '_reindex_axis',\n",
       " '_reindex_indexer',\n",
       " '_reindex_multi',\n",
       " '_reindex_with_indexers',\n",
       " '_repr_data_resource_',\n",
       " '_repr_latex_',\n",
       " '_reset_cache',\n",
       " '_reset_cacher',\n",
       " '_selected_obj',\n",
       " '_selection',\n",
       " '_selection_list',\n",
       " '_selection_name',\n",
       " '_set_as_cached',\n",
       " '_set_axis',\n",
       " '_set_axis_name',\n",
       " '_set_is_copy',\n",
       " '_set_item',\n",
       " '_set_labels',\n",
       " '_set_name',\n",
       " '_set_subtyp',\n",
       " '_set_value',\n",
       " '_set_values',\n",
       " '_set_with',\n",
       " '_set_with_engine',\n",
       " '_setup_axes',\n",
       " '_shallow_copy',\n",
       " '_slice',\n",
       " '_stat_axis',\n",
       " '_stat_axis_name',\n",
       " '_stat_axis_number',\n",
       " '_take',\n",
       " '_to_dict_of_blocks',\n",
       " '_try_aggregate_string_function',\n",
       " '_typ',\n",
       " '_unpickle_series_compat',\n",
       " '_update_inplace',\n",
       " '_validate_dtype',\n",
       " '_values',\n",
       " '_where',\n",
       " '_xs',\n",
       " 'abs',\n",
       " 'add',\n",
       " 'add_prefix',\n",
       " 'add_suffix',\n",
       " 'agg',\n",
       " 'aggregate',\n",
       " 'align',\n",
       " 'all',\n",
       " 'any',\n",
       " 'append',\n",
       " 'apply',\n",
       " 'argmax',\n",
       " 'argmin',\n",
       " 'argsort',\n",
       " 'as_matrix',\n",
       " 'asfreq',\n",
       " 'asobject',\n",
       " 'asof',\n",
       " 'astype',\n",
       " 'at',\n",
       " 'at_time',\n",
       " 'autocorr',\n",
       " 'axes',\n",
       " 'base',\n",
       " 'between',\n",
       " 'between_time',\n",
       " 'bfill',\n",
       " 'bool',\n",
       " 'clip',\n",
       " 'clip_lower',\n",
       " 'clip_upper',\n",
       " 'combine',\n",
       " 'combine_first',\n",
       " 'compound',\n",
       " 'compress',\n",
       " 'copy',\n",
       " 'corr',\n",
       " 'count',\n",
       " 'cov',\n",
       " 'cummax',\n",
       " 'cummin',\n",
       " 'cumprod',\n",
       " 'cumsum',\n",
       " 'data',\n",
       " 'describe',\n",
       " 'diff',\n",
       " 'div',\n",
       " 'divide',\n",
       " 'dot',\n",
       " 'drop',\n",
       " 'drop_duplicates',\n",
       " 'dropna',\n",
       " 'dtype',\n",
       " 'dtypes',\n",
       " 'duplicated',\n",
       " 'empty',\n",
       " 'eq',\n",
       " 'equals',\n",
       " 'ewm',\n",
       " 'expanding',\n",
       " 'factorize',\n",
       " 'ffill',\n",
       " 'fillna',\n",
       " 'filter',\n",
       " 'first',\n",
       " 'first_valid_index',\n",
       " 'flags',\n",
       " 'floordiv',\n",
       " 'from_array',\n",
       " 'ftype',\n",
       " 'ftypes',\n",
       " 'ge',\n",
       " 'get',\n",
       " 'get_dtype_counts',\n",
       " 'get_ftype_counts',\n",
       " 'get_values',\n",
       " 'groupby',\n",
       " 'gt',\n",
       " 'hasnans',\n",
       " 'head',\n",
       " 'hist',\n",
       " 'iat',\n",
       " 'idxmax',\n",
       " 'idxmin',\n",
       " 'iloc',\n",
       " 'imag',\n",
       " 'index',\n",
       " 'infer_objects',\n",
       " 'interpolate',\n",
       " 'is_copy',\n",
       " 'is_monotonic',\n",
       " 'is_monotonic_decreasing',\n",
       " 'is_monotonic_increasing',\n",
       " 'is_unique',\n",
       " 'isin',\n",
       " 'isna',\n",
       " 'isnull',\n",
       " 'item',\n",
       " 'items',\n",
       " 'itemsize',\n",
       " 'iteritems',\n",
       " 'ix',\n",
       " 'keys',\n",
       " 'kurt',\n",
       " 'kurtosis',\n",
       " 'last',\n",
       " 'last_valid_index',\n",
       " 'le',\n",
       " 'loc',\n",
       " 'lt',\n",
       " 'mad',\n",
       " 'map',\n",
       " 'mask',\n",
       " 'max',\n",
       " 'mean',\n",
       " 'median',\n",
       " 'memory_usage',\n",
       " 'min',\n",
       " 'mod',\n",
       " 'mode',\n",
       " 'mul',\n",
       " 'multiply',\n",
       " 'name',\n",
       " 'nbytes',\n",
       " 'ndim',\n",
       " 'ne',\n",
       " 'nlargest',\n",
       " 'nonzero',\n",
       " 'notna',\n",
       " 'notnull',\n",
       " 'nsmallest',\n",
       " 'nunique',\n",
       " 'pct_change',\n",
       " 'pipe',\n",
       " 'plot',\n",
       " 'pop',\n",
       " 'pow',\n",
       " 'prod',\n",
       " 'product',\n",
       " 'ptp',\n",
       " 'put',\n",
       " 'quantile',\n",
       " 'radd',\n",
       " 'rank',\n",
       " 'ravel',\n",
       " 'rdiv',\n",
       " 'real',\n",
       " 'reindex',\n",
       " 'reindex_axis',\n",
       " 'reindex_like',\n",
       " 'rename',\n",
       " 'rename_axis',\n",
       " 'reorder_levels',\n",
       " 'repeat',\n",
       " 'replace',\n",
       " 'resample',\n",
       " 'reset_index',\n",
       " 'rfloordiv',\n",
       " 'rmod',\n",
       " 'rmul',\n",
       " 'rolling',\n",
       " 'round',\n",
       " 'rpow',\n",
       " 'rsub',\n",
       " 'rtruediv',\n",
       " 'sample',\n",
       " 'searchsorted',\n",
       " 'select',\n",
       " 'sem',\n",
       " 'set_axis',\n",
       " 'shape',\n",
       " 'shift',\n",
       " 'size',\n",
       " 'skew',\n",
       " 'slice_shift',\n",
       " 'sort_index',\n",
       " 'sort_values',\n",
       " 'squeeze',\n",
       " 'std',\n",
       " 'str',\n",
       " 'strides',\n",
       " 'sub',\n",
       " 'subtract',\n",
       " 'sum',\n",
       " 'swapaxes',\n",
       " 'swaplevel',\n",
       " 'tail',\n",
       " 'take',\n",
       " 'to_clipboard',\n",
       " 'to_csv',\n",
       " 'to_dense',\n",
       " 'to_dict',\n",
       " 'to_excel',\n",
       " 'to_frame',\n",
       " 'to_hdf',\n",
       " 'to_json',\n",
       " 'to_latex',\n",
       " 'to_msgpack',\n",
       " 'to_period',\n",
       " 'to_pickle',\n",
       " 'to_sparse',\n",
       " 'to_sql',\n",
       " 'to_string',\n",
       " 'to_timestamp',\n",
       " 'to_xarray',\n",
       " 'tolist',\n",
       " 'transform',\n",
       " 'transpose',\n",
       " 'truediv',\n",
       " 'truncate',\n",
       " 'tshift',\n",
       " 'tz_convert',\n",
       " 'tz_localize',\n",
       " 'unique',\n",
       " 'unstack',\n",
       " 'update',\n",
       " 'valid',\n",
       " 'value_counts',\n",
       " 'values',\n",
       " 'var',\n",
       " 'view',\n",
       " 'where',\n",
       " 'xs']"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(s)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "More exciting than the series object and more useful to us, is the dataframe.  Dataframes have rows and columns and dataframes can be created in multiple ways. Also, notice that in the notebook, where code is run interactively, you do not have to use print statements to return output.  However, in pycharm or when running code from scripts, you must still use the print statement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>cat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>dog</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a  b    c\n",
       "0  1  3  cat\n",
       "1  2  4  dog"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = {'a':[1, 2], 'b': (3, 4), 'c': ['cat', 'dog']}\n",
    "pd.DataFrame(df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>a</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c</td>\n",
       "      <td>cat</td>\n",
       "      <td>dog</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0    1    2\n",
       "0  a    1    2\n",
       "1  b    3    4\n",
       "2  c  cat  dog"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = [['a',1,2], ['b', 3,4], ['c', 'cat', 'dog']]\n",
    "pd.DataFrame(df2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a    b    c\n",
       "0  1  4.0  NaN\n",
       "1  2  NaN  3.0"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = [{'a':1, 'b':4}, {'a':2, 'c':3}]\n",
    "pd.DataFrame(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "More important to our uses, we can give Pandas an excel spreadsheet or text file and then play with that data.  And the file i/o is way easier than base python.  The data we are playing with today is compliments of Larsson et al., 2007.  In this study, the authors examined differential expression upon overexpression of a protein in cell culture.  To see what a day in the life of dealing with other people's data is like, we will now ask some questions of this dataset.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice the last column, has empty space or missing data.  Pandas will automatically deal with that."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![alt text](Screen Shot 2016-09-28 at 9.24.25 AM.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#this reads the data into a dataframe object, using a tab separated character.  The index_col argument is optional\n",
    "# and sets the first column in the data as rownames which is similar to how R works.\n",
    "genelist = pd.read_table('Supplement_table02.txt', sep='\\t', index_col=0)\n",
    "# genelist_excel = pd.read_excel('Supplement_table02.xlsx', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function read_table in module pandas.io.parsers:\n",
      "\n",
      "read_table(filepath_or_buffer, sep='\\t', delimiter=None, header='infer', names=None, index_col=None, usecols=None, squeeze=False, prefix=None, mangle_dupe_cols=True, dtype=None, engine=None, converters=None, true_values=None, false_values=None, skipinitialspace=False, skiprows=None, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, skip_blank_lines=True, parse_dates=False, infer_datetime_format=False, keep_date_col=False, date_parser=None, dayfirst=False, iterator=False, chunksize=None, compression='infer', thousands=None, decimal=b'.', lineterminator=None, quotechar='\"', quoting=0, escapechar=None, comment=None, encoding=None, dialect=None, tupleize_cols=None, error_bad_lines=True, warn_bad_lines=True, skipfooter=0, skip_footer=0, doublequote=True, delim_whitespace=False, as_recarray=None, compact_ints=None, use_unsigned=None, low_memory=True, buffer_lines=None, memory_map=False, float_precision=None)\n",
      "    Read general delimited file into DataFrame\n",
      "    \n",
      "    Also supports optionally iterating or breaking of the file\n",
      "    into chunks.\n",
      "    \n",
      "    Additional help can be found in the `online docs for IO Tools\n",
      "    <http://pandas.pydata.org/pandas-docs/stable/io.html>`_.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    filepath_or_buffer : str, pathlib.Path, py._path.local.LocalPath or any object with a read() method (such as a file handle or StringIO)\n",
      "        The string could be a URL. Valid URL schemes include http, ftp, s3, and\n",
      "        file. For file URLs, a host is expected. For instance, a local file could\n",
      "        be file ://localhost/path/to/table.csv\n",
      "    sep : str, default \\t (tab-stop)\n",
      "        Delimiter to use. If sep is None, the C engine cannot automatically detect\n",
      "        the separator, but the Python parsing engine can, meaning the latter will\n",
      "        be used and automatically detect the separator by Python's builtin sniffer\n",
      "        tool, ``csv.Sniffer``. In addition, separators longer than 1 character and\n",
      "        different from ``'\\s+'`` will be interpreted as regular expressions and\n",
      "        will also force the use of the Python parsing engine. Note that regex\n",
      "        delimiters are prone to ignoring quoted data. Regex example: ``'\\r\\t'``\n",
      "    delimiter : str, default ``None``\n",
      "        Alternative argument name for sep.\n",
      "    delim_whitespace : boolean, default False\n",
      "        Specifies whether or not whitespace (e.g. ``' '`` or ``'    '``) will be\n",
      "        used as the sep. Equivalent to setting ``sep='\\s+'``. If this option\n",
      "        is set to True, nothing should be passed in for the ``delimiter``\n",
      "        parameter.\n",
      "    \n",
      "        .. versionadded:: 0.18.1 support for the Python parser.\n",
      "    \n",
      "    header : int or list of ints, default 'infer'\n",
      "        Row number(s) to use as the column names, and the start of the data.\n",
      "        Default behavior is as if set to 0 if no ``names`` passed, otherwise\n",
      "        ``None``. Explicitly pass ``header=0`` to be able to replace existing\n",
      "        names. The header can be a list of integers that specify row locations for\n",
      "        a multi-index on the columns e.g. [0,1,3]. Intervening rows that are not\n",
      "        specified will be skipped (e.g. 2 in this example is skipped). Note that\n",
      "        this parameter ignores commented lines and empty lines if\n",
      "        ``skip_blank_lines=True``, so header=0 denotes the first line of data\n",
      "        rather than the first line of the file.\n",
      "    names : array-like, default None\n",
      "        List of column names to use. If file contains no header row, then you\n",
      "        should explicitly pass header=None. Duplicates in this list will cause\n",
      "        a ``UserWarning`` to be issued.\n",
      "    index_col : int or sequence or False, default None\n",
      "        Column to use as the row labels of the DataFrame. If a sequence is given, a\n",
      "        MultiIndex is used. If you have a malformed file with delimiters at the end\n",
      "        of each line, you might consider index_col=False to force pandas to _not_\n",
      "        use the first column as the index (row names)\n",
      "    usecols : array-like or callable, default None\n",
      "        Return a subset of the columns. If array-like, all elements must either\n",
      "        be positional (i.e. integer indices into the document columns) or strings\n",
      "        that correspond to column names provided either by the user in `names` or\n",
      "        inferred from the document header row(s). For example, a valid array-like\n",
      "        `usecols` parameter would be [0, 1, 2] or ['foo', 'bar', 'baz'].\n",
      "    \n",
      "        If callable, the callable function will be evaluated against the column\n",
      "        names, returning names where the callable function evaluates to True. An\n",
      "        example of a valid callable argument would be ``lambda x: x.upper() in\n",
      "        ['AAA', 'BBB', 'DDD']``. Using this parameter results in much faster\n",
      "        parsing time and lower memory usage.\n",
      "    as_recarray : boolean, default False\n",
      "        .. deprecated:: 0.19.0\n",
      "           Please call `pd.read_csv(...).to_records()` instead.\n",
      "    \n",
      "        Return a NumPy recarray instead of a DataFrame after parsing the data.\n",
      "        If set to True, this option takes precedence over the `squeeze` parameter.\n",
      "        In addition, as row indices are not available in such a format, the\n",
      "        `index_col` parameter will be ignored.\n",
      "    squeeze : boolean, default False\n",
      "        If the parsed data only contains one column then return a Series\n",
      "    prefix : str, default None\n",
      "        Prefix to add to column numbers when no header, e.g. 'X' for X0, X1, ...\n",
      "    mangle_dupe_cols : boolean, default True\n",
      "        Duplicate columns will be specified as 'X.0'...'X.N', rather than\n",
      "        'X'...'X'. Passing in False will cause data to be overwritten if there\n",
      "        are duplicate names in the columns.\n",
      "    dtype : Type name or dict of column -> type, default None\n",
      "        Data type for data or columns. E.g. {'a': np.float64, 'b': np.int32}\n",
      "        Use `str` or `object` to preserve and not interpret dtype.\n",
      "        If converters are specified, they will be applied INSTEAD\n",
      "        of dtype conversion.\n",
      "    engine : {'c', 'python'}, optional\n",
      "        Parser engine to use. The C engine is faster while the python engine is\n",
      "        currently more feature-complete.\n",
      "    converters : dict, default None\n",
      "        Dict of functions for converting values in certain columns. Keys can either\n",
      "        be integers or column labels\n",
      "    true_values : list, default None\n",
      "        Values to consider as True\n",
      "    false_values : list, default None\n",
      "        Values to consider as False\n",
      "    skipinitialspace : boolean, default False\n",
      "        Skip spaces after delimiter.\n",
      "    skiprows : list-like or integer or callable, default None\n",
      "        Line numbers to skip (0-indexed) or number of lines to skip (int)\n",
      "        at the start of the file.\n",
      "    \n",
      "        If callable, the callable function will be evaluated against the row\n",
      "        indices, returning True if the row should be skipped and False otherwise.\n",
      "        An example of a valid callable argument would be ``lambda x: x in [0, 2]``.\n",
      "    skipfooter : int, default 0\n",
      "        Number of lines at bottom of file to skip (Unsupported with engine='c')\n",
      "    skip_footer : int, default 0\n",
      "        .. deprecated:: 0.19.0\n",
      "           Use the `skipfooter` parameter instead, as they are identical\n",
      "    nrows : int, default None\n",
      "        Number of rows of file to read. Useful for reading pieces of large files\n",
      "    na_values : scalar, str, list-like, or dict, default None\n",
      "        Additional strings to recognize as NA/NaN. If dict passed, specific\n",
      "        per-column NA values.  By default the following values are interpreted as\n",
      "        NaN: '', '#N/A', '#N/A N/A', '#NA', '-1.#IND', '-1.#QNAN', '-NaN', '-nan',\n",
      "        '1.#IND', '1.#QNAN', 'N/A', 'NA', 'NULL', 'NaN', 'n/a', 'nan',\n",
      "        'null'.\n",
      "    keep_default_na : bool, default True\n",
      "        If na_values are specified and keep_default_na is False the default NaN\n",
      "        values are overridden, otherwise they're appended to.\n",
      "    na_filter : boolean, default True\n",
      "        Detect missing value markers (empty strings and the value of na_values). In\n",
      "        data without any NAs, passing na_filter=False can improve the performance\n",
      "        of reading a large file\n",
      "    verbose : boolean, default False\n",
      "        Indicate number of NA values placed in non-numeric columns\n",
      "    skip_blank_lines : boolean, default True\n",
      "        If True, skip over blank lines rather than interpreting as NaN values\n",
      "    parse_dates : boolean or list of ints or names or list of lists or dict, default False\n",
      "    \n",
      "        * boolean. If True -> try parsing the index.\n",
      "        * list of ints or names. e.g. If [1, 2, 3] -> try parsing columns 1, 2, 3\n",
      "          each as a separate date column.\n",
      "        * list of lists. e.g.  If [[1, 3]] -> combine columns 1 and 3 and parse as\n",
      "          a single date column.\n",
      "        * dict, e.g. {'foo' : [1, 3]} -> parse columns 1, 3 as date and call result\n",
      "          'foo'\n",
      "    \n",
      "        If a column or index contains an unparseable date, the entire column or\n",
      "        index will be returned unaltered as an object data type. For non-standard\n",
      "        datetime parsing, use ``pd.to_datetime`` after ``pd.read_csv``\n",
      "    \n",
      "        Note: A fast-path exists for iso8601-formatted dates.\n",
      "    infer_datetime_format : boolean, default False\n",
      "        If True and `parse_dates` is enabled, pandas will attempt to infer the\n",
      "        format of the datetime strings in the columns, and if it can be inferred,\n",
      "        switch to a faster method of parsing them. In some cases this can increase\n",
      "        the parsing speed by 5-10x.\n",
      "    keep_date_col : boolean, default False\n",
      "        If True and `parse_dates` specifies combining multiple columns then\n",
      "        keep the original columns.\n",
      "    date_parser : function, default None\n",
      "        Function to use for converting a sequence of string columns to an array of\n",
      "        datetime instances. The default uses ``dateutil.parser.parser`` to do the\n",
      "        conversion. Pandas will try to call `date_parser` in three different ways,\n",
      "        advancing to the next if an exception occurs: 1) Pass one or more arrays\n",
      "        (as defined by `parse_dates`) as arguments; 2) concatenate (row-wise) the\n",
      "        string values from the columns defined by `parse_dates` into a single array\n",
      "        and pass that; and 3) call `date_parser` once for each row using one or\n",
      "        more strings (corresponding to the columns defined by `parse_dates`) as\n",
      "        arguments.\n",
      "    dayfirst : boolean, default False\n",
      "        DD/MM format dates, international and European format\n",
      "    iterator : boolean, default False\n",
      "        Return TextFileReader object for iteration or getting chunks with\n",
      "        ``get_chunk()``.\n",
      "    chunksize : int, default None\n",
      "        Return TextFileReader object for iteration.\n",
      "        See the `IO Tools docs\n",
      "        <http://pandas.pydata.org/pandas-docs/stable/io.html#io-chunking>`_\n",
      "        for more information on ``iterator`` and ``chunksize``.\n",
      "    compression : {'infer', 'gzip', 'bz2', 'zip', 'xz', None}, default 'infer'\n",
      "        For on-the-fly decompression of on-disk data. If 'infer' and\n",
      "        `filepath_or_buffer` is path-like, then detect compression from the\n",
      "        following extensions: '.gz', '.bz2', '.zip', or '.xz' (otherwise no\n",
      "        decompression). If using 'zip', the ZIP file must contain only one data\n",
      "        file to be read in. Set to None for no decompression.\n",
      "    \n",
      "        .. versionadded:: 0.18.1 support for 'zip' and 'xz' compression.\n",
      "    \n",
      "    thousands : str, default None\n",
      "        Thousands separator\n",
      "    decimal : str, default '.'\n",
      "        Character to recognize as decimal point (e.g. use ',' for European data).\n",
      "    float_precision : string, default None\n",
      "        Specifies which converter the C engine should use for floating-point\n",
      "        values. The options are `None` for the ordinary converter,\n",
      "        `high` for the high-precision converter, and `round_trip` for the\n",
      "        round-trip converter.\n",
      "    lineterminator : str (length 1), default None\n",
      "        Character to break file into lines. Only valid with C parser.\n",
      "    quotechar : str (length 1), optional\n",
      "        The character used to denote the start and end of a quoted item. Quoted\n",
      "        items can include the delimiter and it will be ignored.\n",
      "    quoting : int or csv.QUOTE_* instance, default 0\n",
      "        Control field quoting behavior per ``csv.QUOTE_*`` constants. Use one of\n",
      "        QUOTE_MINIMAL (0), QUOTE_ALL (1), QUOTE_NONNUMERIC (2) or QUOTE_NONE (3).\n",
      "    doublequote : boolean, default ``True``\n",
      "       When quotechar is specified and quoting is not ``QUOTE_NONE``, indicate\n",
      "       whether or not to interpret two consecutive quotechar elements INSIDE a\n",
      "       field as a single ``quotechar`` element.\n",
      "    escapechar : str (length 1), default None\n",
      "        One-character string used to escape delimiter when quoting is QUOTE_NONE.\n",
      "    comment : str, default None\n",
      "        Indicates remainder of line should not be parsed. If found at the beginning\n",
      "        of a line, the line will be ignored altogether. This parameter must be a\n",
      "        single character. Like empty lines (as long as ``skip_blank_lines=True``),\n",
      "        fully commented lines are ignored by the parameter `header` but not by\n",
      "        `skiprows`. For example, if comment='#', parsing '#empty\\na,b,c\\n1,2,3'\n",
      "        with `header=0` will result in 'a,b,c' being\n",
      "        treated as the header.\n",
      "    encoding : str, default None\n",
      "        Encoding to use for UTF when reading/writing (ex. 'utf-8'). `List of Python\n",
      "        standard encodings\n",
      "        <https://docs.python.org/3/library/codecs.html#standard-encodings>`_\n",
      "    dialect : str or csv.Dialect instance, default None\n",
      "        If provided, this parameter will override values (default or not) for the\n",
      "        following parameters: `delimiter`, `doublequote`, `escapechar`,\n",
      "        `skipinitialspace`, `quotechar`, and `quoting`. If it is necessary to\n",
      "        override values, a ParserWarning will be issued. See csv.Dialect\n",
      "        documentation for more details.\n",
      "    tupleize_cols : boolean, default False\n",
      "        .. deprecated:: 0.21.0\n",
      "           This argument will be removed and will always convert to MultiIndex\n",
      "    \n",
      "        Leave a list of tuples on columns as is (default is to convert to\n",
      "        a MultiIndex on the columns)\n",
      "    error_bad_lines : boolean, default True\n",
      "        Lines with too many fields (e.g. a csv line with too many commas) will by\n",
      "        default cause an exception to be raised, and no DataFrame will be returned.\n",
      "        If False, then these \"bad lines\" will dropped from the DataFrame that is\n",
      "        returned.\n",
      "    warn_bad_lines : boolean, default True\n",
      "        If error_bad_lines is False, and warn_bad_lines is True, a warning for each\n",
      "        \"bad line\" will be output.\n",
      "    low_memory : boolean, default True\n",
      "        Internally process the file in chunks, resulting in lower memory use\n",
      "        while parsing, but possibly mixed type inference.  To ensure no mixed\n",
      "        types either set False, or specify the type with the `dtype` parameter.\n",
      "        Note that the entire file is read into a single DataFrame regardless,\n",
      "        use the `chunksize` or `iterator` parameter to return the data in chunks.\n",
      "        (Only valid with C parser)\n",
      "    buffer_lines : int, default None\n",
      "        .. deprecated:: 0.19.0\n",
      "           This argument is not respected by the parser\n",
      "    compact_ints : boolean, default False\n",
      "        .. deprecated:: 0.19.0\n",
      "           Argument moved to ``pd.to_numeric``\n",
      "    \n",
      "        If compact_ints is True, then for any column that is of integer dtype,\n",
      "        the parser will attempt to cast it as the smallest integer dtype possible,\n",
      "        either signed or unsigned depending on the specification from the\n",
      "        `use_unsigned` parameter.\n",
      "    use_unsigned : boolean, default False\n",
      "        .. deprecated:: 0.19.0\n",
      "           Argument moved to ``pd.to_numeric``\n",
      "    \n",
      "        If integer columns are being compacted (i.e. `compact_ints=True`), specify\n",
      "        whether the column should be compacted to the smallest signed or unsigned\n",
      "        integer dtype.\n",
      "    memory_map : boolean, default False\n",
      "        If a filepath is provided for `filepath_or_buffer`, map the file object\n",
      "        directly onto memory and access the data directly from there. Using this\n",
      "        option can improve performance because there is no longer any I/O overhead.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    result : DataFrame or TextParser\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pd.read_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viewing the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's view the data frame.  You can see that Pandas filled in missing values with NaN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_130786</th>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>2.176788</td>\n",
       "      <td>1.998233</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_015665</th>\n",
       "      <td>8086.0</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>achalasia, adrenocortical insufficiency, alacr...</td>\n",
       "      <td>1.952652</td>\n",
       "      <td>3.753678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001087</th>\n",
       "      <td>14.0</td>\n",
       "      <td>AAMP</td>\n",
       "      <td>angio-associated, migratory cell protein</td>\n",
       "      <td>2.808783</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_025267</th>\n",
       "      <td>80755.0</td>\n",
       "      <td>AARSD1</td>\n",
       "      <td>alanyl-tRNA synthetase domain containing 1</td>\n",
       "      <td>1.962436</td>\n",
       "      <td>9.865391</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_012089</th>\n",
       "      <td>23456.0</td>\n",
       "      <td>ABCB10</td>\n",
       "      <td>ATP-binding cassette, sub-family B (MDR/TAP), ...</td>\n",
       "      <td>0.225564</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-4.433328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_007189</th>\n",
       "      <td>10061.0</td>\n",
       "      <td>ABCF2</td>\n",
       "      <td>ATP-binding cassette, sub-family F (GCN20), me...</td>\n",
       "      <td>2.226917</td>\n",
       "      <td>8.039531</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_005692</th>\n",
       "      <td>10061.0</td>\n",
       "      <td>ABCF2</td>\n",
       "      <td>ATP-binding cassette, sub-family F (GCN20), me...</td>\n",
       "      <td>2.661841</td>\n",
       "      <td>8.727441</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_015407</th>\n",
       "      <td>25864.0</td>\n",
       "      <td>ABHD14A</td>\n",
       "      <td>abhydrolase domain containing 14A</td>\n",
       "      <td>0.462046</td>\n",
       "      <td>9.428327</td>\n",
       "      <td>-2.164288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_020676</th>\n",
       "      <td>57406.0</td>\n",
       "      <td>ABHD6</td>\n",
       "      <td>abhydrolase domain containing 6</td>\n",
       "      <td>0.394502</td>\n",
       "      <td>8.039531</td>\n",
       "      <td>-2.534839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_005759</th>\n",
       "      <td>10152.0</td>\n",
       "      <td>ABI2</td>\n",
       "      <td>abl interactor 2</td>\n",
       "      <td>1.621564</td>\n",
       "      <td>7.653744</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_014945</th>\n",
       "      <td>22885.0</td>\n",
       "      <td>ABLIM3</td>\n",
       "      <td>actin binding LIM protein family, member 3</td>\n",
       "      <td>2.205244</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_013375</th>\n",
       "      <td>29777.0</td>\n",
       "      <td>ABT1</td>\n",
       "      <td>activator of basal transcription 1</td>\n",
       "      <td>3.284837</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_005891</th>\n",
       "      <td>39.0</td>\n",
       "      <td>ACAT2</td>\n",
       "      <td>acetyl-Coenzyme A acetyltransferase 2 (acetoac...</td>\n",
       "      <td>1.664480</td>\n",
       "      <td>8.593663</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_022914</th>\n",
       "      <td>65057.0</td>\n",
       "      <td>ACD</td>\n",
       "      <td>adrenocortical dysplasia homolog (mouse)</td>\n",
       "      <td>1.766179</td>\n",
       "      <td>5.763981</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_181865</th>\n",
       "      <td>11332.0</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>acyl-CoA thioesterase 7</td>\n",
       "      <td>1.953817</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_181866</th>\n",
       "      <td>11332.0</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>acyl-CoA thioesterase 7</td>\n",
       "      <td>1.953817</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_181864</th>\n",
       "      <td>11332.0</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>acyl-CoA thioesterase 7</td>\n",
       "      <td>1.953817</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_007274</th>\n",
       "      <td>11332.0</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>acyl-CoA thioesterase 7</td>\n",
       "      <td>1.953817</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_004300</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.790544</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_177554</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.288103</td>\n",
       "      <td>1.219488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_007099</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.288103</td>\n",
       "      <td>1.219488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_203372</th>\n",
       "      <td>2181.0</td>\n",
       "      <td>ACSL3</td>\n",
       "      <td>acyl-CoA synthetase long-chain family member 3</td>\n",
       "      <td>0.254533</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-3.928758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_004457</th>\n",
       "      <td>2181.0</td>\n",
       "      <td>ACSL3</td>\n",
       "      <td>acyl-CoA synthetase long-chain family member 3</td>\n",
       "      <td>0.254533</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-3.928758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_005736</th>\n",
       "      <td>10121.0</td>\n",
       "      <td>ACTR1A</td>\n",
       "      <td>ARP1 actin-related protein 1 homolog A, centra...</td>\n",
       "      <td>2.088440</td>\n",
       "      <td>2.473819</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_005735</th>\n",
       "      <td>10120.0</td>\n",
       "      <td>ACTR1B</td>\n",
       "      <td>ARP1 actin-related protein 1 homolog B, centra...</td>\n",
       "      <td>1.846134</td>\n",
       "      <td>8.039531</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_020445</th>\n",
       "      <td>57180.0</td>\n",
       "      <td>ACTR3B</td>\n",
       "      <td>ARP3 actin-related protein 3 homolog B (yeast)</td>\n",
       "      <td>1.622898</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000666</th>\n",
       "      <td>95.0</td>\n",
       "      <td>ACY1</td>\n",
       "      <td>aminoacylase 1</td>\n",
       "      <td>2.271014</td>\n",
       "      <td>5.370618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001107</th>\n",
       "      <td>97.0</td>\n",
       "      <td>ACYP1</td>\n",
       "      <td>acylphosphatase 1, erythrocyte (common) type</td>\n",
       "      <td>1.480483</td>\n",
       "      <td>9.865391</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_138448</th>\n",
       "      <td>98.0</td>\n",
       "      <td>ACYP2</td>\n",
       "      <td>acylphosphatase 2, muscle type</td>\n",
       "      <td>2.028069</td>\n",
       "      <td>5.763981</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000022</th>\n",
       "      <td>100.0</td>\n",
       "      <td>ADA</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>3.166759</td>\n",
       "      <td>1.021824</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_021906</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.441336</td>\n",
       "      <td>3.753678</td>\n",
       "      <td>-2.265846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_145323</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.266313</td>\n",
       "      <td>2.016183</td>\n",
       "      <td>-3.754986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_039627</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.445723</td>\n",
       "      <td>9.428327</td>\n",
       "      <td>-2.243547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_016496</th>\n",
       "      <td>51257.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>membrane-associated ring finger (C3HC4) 2</td>\n",
       "      <td>2.220592</td>\n",
       "      <td>4.319905</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_148968</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.575574</td>\n",
       "      <td>6.349791</td>\n",
       "      <td>-1.737395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_211529</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.801124</td>\n",
       "      <td>4.319905</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_372143</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.019991</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_148973</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.575574</td>\n",
       "      <td>6.349791</td>\n",
       "      <td>-1.737395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_148972</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.575574</td>\n",
       "      <td>6.349791</td>\n",
       "      <td>-1.737395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_002097</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.924556</td>\n",
       "      <td>6.599963</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_015022</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.424889</td>\n",
       "      <td>8.866014</td>\n",
       "      <td>-2.353558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_379250</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.057951</td>\n",
       "      <td>2.103816</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-HUMRGE/M10098_5_at</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.350925</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-2.849608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_374065</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.486665</td>\n",
       "      <td>2.491259</td>\n",
       "      <td>-2.054800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_379694</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.999065</td>\n",
       "      <td>3.797649</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_371848</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.970303</td>\n",
       "      <td>2.601228</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XR_000223</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.038178</td>\n",
       "      <td>4.934189</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NR_002229</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.618970</td>\n",
       "      <td>8.593663</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_371838</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.466004</td>\n",
       "      <td>9.078295</td>\n",
       "      <td>-2.145902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_499128</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.022438</td>\n",
       "      <td>1.219488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_372719</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.349097</td>\n",
       "      <td>2.016183</td>\n",
       "      <td>-2.864532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_498218</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.777962</td>\n",
       "      <td>7.653744</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_016572</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.102908</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_379274</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.172858</td>\n",
       "      <td>1.939343</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_114303</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.446474</td>\n",
       "      <td>2.590323</td>\n",
       "      <td>-2.239770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_148969</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.575574</td>\n",
       "      <td>6.349791</td>\n",
       "      <td>-1.737395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_496301</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.374709</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.668740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_181862</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.953817</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_145646</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.476765</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>-2.097469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_153032</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.087563</td>\n",
       "      <td>5.763981</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3085 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Geneid Gene symbol  \\\n",
       "NM_130786                    1.0        A1BG   \n",
       "NM_015665                 8086.0        AAAS   \n",
       "NM_001087                   14.0        AAMP   \n",
       "NM_025267                80755.0      AARSD1   \n",
       "NM_012089                23456.0      ABCB10   \n",
       "NM_007189                10061.0       ABCF2   \n",
       "NM_005692                10061.0       ABCF2   \n",
       "NM_015407                25864.0     ABHD14A   \n",
       "NM_020676                57406.0       ABHD6   \n",
       "NM_005759                10152.0        ABI2   \n",
       "NM_014945                22885.0      ABLIM3   \n",
       "NM_013375                29777.0        ABT1   \n",
       "NM_005891                   39.0       ACAT2   \n",
       "NM_022914                65057.0         ACD   \n",
       "NM_181865                11332.0       ACOT7   \n",
       "NM_181866                11332.0       ACOT7   \n",
       "NM_181864                11332.0       ACOT7   \n",
       "NM_007274                11332.0       ACOT7   \n",
       "NM_004300                   52.0        ACP1   \n",
       "NM_177554                   52.0        ACP1   \n",
       "NM_007099                   52.0        ACP1   \n",
       "NM_203372                 2181.0       ACSL3   \n",
       "NM_004457                 2181.0       ACSL3   \n",
       "NM_005736                10121.0      ACTR1A   \n",
       "NM_005735                10120.0      ACTR1B   \n",
       "NM_020445                57180.0      ACTR3B   \n",
       "NM_000666                   95.0        ACY1   \n",
       "NM_001107                   97.0       ACYP1   \n",
       "NM_138448                   98.0       ACYP2   \n",
       "NM_000022                  100.0         ADA   \n",
       "...                          ...         ...   \n",
       "NM_021906                    NaN         NaN   \n",
       "NM_145323                    NaN         NaN   \n",
       "XM_039627                    NaN         NaN   \n",
       "NM_016496                51257.0         NaN   \n",
       "NM_148968                    NaN         NaN   \n",
       "XM_211529                    NaN         NaN   \n",
       "XM_372143                    NaN         NaN   \n",
       "NM_148973                    NaN         NaN   \n",
       "NM_148972                    NaN         NaN   \n",
       "NM_002097                    NaN         NaN   \n",
       "NM_015022                    NaN         NaN   \n",
       "XM_379250                    NaN         NaN   \n",
       "AFFX-HUMRGE/M10098_5_at      NaN         NaN   \n",
       "XM_374065                    NaN         NaN   \n",
       "XM_379694                    NaN         NaN   \n",
       "XM_371848                    NaN         NaN   \n",
       "XR_000223                    NaN         NaN   \n",
       "NR_002229                    NaN         NaN   \n",
       "XM_371838                    NaN         NaN   \n",
       "XM_499128                    NaN         NaN   \n",
       "XM_372719                    NaN         NaN   \n",
       "XM_498218                    NaN         NaN   \n",
       "NM_016572                    NaN         NaN   \n",
       "XM_379274                    NaN         NaN   \n",
       "XM_114303                    NaN         NaN   \n",
       "NM_148969                    NaN         NaN   \n",
       "XM_496301                    NaN         NaN   \n",
       "NM_181862                    NaN         NaN   \n",
       "NM_145646                    NaN         NaN   \n",
       "NM_153032                    NaN         NaN   \n",
       "\n",
       "                                                                 Gene name  \\\n",
       "NM_130786                                           alpha-1-B glycoprotein   \n",
       "NM_015665                achalasia, adrenocortical insufficiency, alacr...   \n",
       "NM_001087                         angio-associated, migratory cell protein   \n",
       "NM_025267                       alanyl-tRNA synthetase domain containing 1   \n",
       "NM_012089                ATP-binding cassette, sub-family B (MDR/TAP), ...   \n",
       "NM_007189                ATP-binding cassette, sub-family F (GCN20), me...   \n",
       "NM_005692                ATP-binding cassette, sub-family F (GCN20), me...   \n",
       "NM_015407                                abhydrolase domain containing 14A   \n",
       "NM_020676                                  abhydrolase domain containing 6   \n",
       "NM_005759                                                 abl interactor 2   \n",
       "NM_014945                       actin binding LIM protein family, member 3   \n",
       "NM_013375                               activator of basal transcription 1   \n",
       "NM_005891                acetyl-Coenzyme A acetyltransferase 2 (acetoac...   \n",
       "NM_022914                         adrenocortical dysplasia homolog (mouse)   \n",
       "NM_181865                                          acyl-CoA thioesterase 7   \n",
       "NM_181866                                          acyl-CoA thioesterase 7   \n",
       "NM_181864                                          acyl-CoA thioesterase 7   \n",
       "NM_007274                                          acyl-CoA thioesterase 7   \n",
       "NM_004300                                      acid phosphatase 1, soluble   \n",
       "NM_177554                                      acid phosphatase 1, soluble   \n",
       "NM_007099                                      acid phosphatase 1, soluble   \n",
       "NM_203372                   acyl-CoA synthetase long-chain family member 3   \n",
       "NM_004457                   acyl-CoA synthetase long-chain family member 3   \n",
       "NM_005736                ARP1 actin-related protein 1 homolog A, centra...   \n",
       "NM_005735                ARP1 actin-related protein 1 homolog B, centra...   \n",
       "NM_020445                   ARP3 actin-related protein 3 homolog B (yeast)   \n",
       "NM_000666                                                   aminoacylase 1   \n",
       "NM_001107                     acylphosphatase 1, erythrocyte (common) type   \n",
       "NM_138448                                   acylphosphatase 2, muscle type   \n",
       "NM_000022                                              adenosine deaminase   \n",
       "...                                                                    ...   \n",
       "NM_021906                                                              NaN   \n",
       "NM_145323                                                              NaN   \n",
       "XM_039627                                                              NaN   \n",
       "NM_016496                        membrane-associated ring finger (C3HC4) 2   \n",
       "NM_148968                                                              NaN   \n",
       "XM_211529                                                              NaN   \n",
       "XM_372143                                                              NaN   \n",
       "NM_148973                                                              NaN   \n",
       "NM_148972                                                              NaN   \n",
       "NM_002097                                                              NaN   \n",
       "NM_015022                                                              NaN   \n",
       "XM_379250                                                              NaN   \n",
       "AFFX-HUMRGE/M10098_5_at                                                NaN   \n",
       "XM_374065                                                              NaN   \n",
       "XM_379694                                                              NaN   \n",
       "XM_371848                                                              NaN   \n",
       "XR_000223                                                              NaN   \n",
       "NR_002229                                                              NaN   \n",
       "XM_371838                                                              NaN   \n",
       "XM_499128                                                              NaN   \n",
       "XM_372719                                                              NaN   \n",
       "XM_498218                                                              NaN   \n",
       "NM_016572                                                              NaN   \n",
       "XM_379274                                                              NaN   \n",
       "XM_114303                                                              NaN   \n",
       "NM_148969                                                              NaN   \n",
       "XM_496301                                                              NaN   \n",
       "NM_181862                                                              NaN   \n",
       "NM_145646                                                              NaN   \n",
       "NM_153032                                                              NaN   \n",
       "\n",
       "                         Fold Change  q-value (%)  Negative Fold Change  \n",
       "NM_130786                   2.176788     1.998233                   NaN  \n",
       "NM_015665                   1.952652     3.753678                   NaN  \n",
       "NM_001087                   2.808783     0.637618                   NaN  \n",
       "NM_025267                   1.962436     9.865391                   NaN  \n",
       "NM_012089                   0.225564     1.707905             -4.433328  \n",
       "NM_007189                   2.226917     8.039531                   NaN  \n",
       "NM_005692                   2.661841     8.727441                   NaN  \n",
       "NM_015407                   0.462046     9.428327             -2.164288  \n",
       "NM_020676                   0.394502     8.039531             -2.534839  \n",
       "NM_005759                   1.621564     7.653744                   NaN  \n",
       "NM_014945                   2.205244     4.611046                   NaN  \n",
       "NM_013375                   3.284837     0.637618                   NaN  \n",
       "NM_005891                   1.664480     8.593663                   NaN  \n",
       "NM_022914                   1.766179     5.763981                   NaN  \n",
       "NM_181865                   1.953817     8.270250                   NaN  \n",
       "NM_181866                   1.953817     8.270250                   NaN  \n",
       "NM_181864                   1.953817     8.270250                   NaN  \n",
       "NM_007274                   1.953817     8.270250                   NaN  \n",
       "NM_004300                   2.790544     0.637618                   NaN  \n",
       "NM_177554                   2.288103     1.219488                   NaN  \n",
       "NM_007099                   2.288103     1.219488                   NaN  \n",
       "NM_203372                   0.254533     1.707905             -3.928758  \n",
       "NM_004457                   0.254533     1.707905             -3.928758  \n",
       "NM_005736                   2.088440     2.473819                   NaN  \n",
       "NM_005735                   1.846134     8.039531                   NaN  \n",
       "NM_020445                   1.622898     8.270250                   NaN  \n",
       "NM_000666                   2.271014     5.370618                   NaN  \n",
       "NM_001107                   1.480483     9.865391                   NaN  \n",
       "NM_138448                   2.028069     5.763981                   NaN  \n",
       "NM_000022                   3.166759     1.021824                   NaN  \n",
       "...                              ...          ...                   ...  \n",
       "NM_021906                   0.441336     3.753678             -2.265846  \n",
       "NM_145323                   0.266313     2.016183             -3.754986  \n",
       "XM_039627                   0.445723     9.428327             -2.243547  \n",
       "NM_016496                   2.220592     4.319905                   NaN  \n",
       "NM_148968                   0.575574     6.349791             -1.737395  \n",
       "XM_211529                   1.801124     4.319905                   NaN  \n",
       "XM_372143                   3.019991     0.637618                   NaN  \n",
       "NM_148973                   0.575574     6.349791             -1.737395  \n",
       "NM_148972                   0.575574     6.349791             -1.737395  \n",
       "NM_002097                   1.924556     6.599963                   NaN  \n",
       "NM_015022                   0.424889     8.866014             -2.353558  \n",
       "XM_379250                   3.057951     2.103816                   NaN  \n",
       "AFFX-HUMRGE/M10098_5_at     0.350925     1.707905             -2.849608  \n",
       "XM_374065                   0.486665     2.491259             -2.054800  \n",
       "XM_379694                   1.999065     3.797649                   NaN  \n",
       "XM_371848                   1.970303     2.601228                   NaN  \n",
       "XR_000223                   2.038178     4.934189                   NaN  \n",
       "NR_002229                   1.618970     8.593663                   NaN  \n",
       "XM_371838                   0.466004     9.078295             -2.145902  \n",
       "XM_499128                   2.022438     1.219488                   NaN  \n",
       "XM_372719                   0.349097     2.016183             -2.864532  \n",
       "XM_498218                   1.777962     7.653744                   NaN  \n",
       "NM_016572                   2.102908     4.611046                   NaN  \n",
       "XM_379274                   2.172858     1.939343                   NaN  \n",
       "XM_114303                   0.446474     2.590323             -2.239770  \n",
       "NM_148969                   0.575574     6.349791             -1.737395  \n",
       "XM_496301                   0.374709     3.399167             -2.668740  \n",
       "NM_181862                   1.953817     8.270250                   NaN  \n",
       "NM_145646                   0.476765     8.270250             -2.097469  \n",
       "NM_153032                   2.087563     5.763981                   NaN  \n",
       "\n",
       "[3085 rows x 6 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_130786</th>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>2.176788</td>\n",
       "      <td>1.998233</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_015665</th>\n",
       "      <td>8086.0</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>achalasia, adrenocortical insufficiency, alacr...</td>\n",
       "      <td>1.952652</td>\n",
       "      <td>3.753678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001087</th>\n",
       "      <td>14.0</td>\n",
       "      <td>AAMP</td>\n",
       "      <td>angio-associated, migratory cell protein</td>\n",
       "      <td>2.808783</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_025267</th>\n",
       "      <td>80755.0</td>\n",
       "      <td>AARSD1</td>\n",
       "      <td>alanyl-tRNA synthetase domain containing 1</td>\n",
       "      <td>1.962436</td>\n",
       "      <td>9.865391</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_012089</th>\n",
       "      <td>23456.0</td>\n",
       "      <td>ABCB10</td>\n",
       "      <td>ATP-binding cassette, sub-family B (MDR/TAP), ...</td>\n",
       "      <td>0.225564</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-4.433328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Geneid Gene symbol  \\\n",
       "NM_130786      1.0        A1BG   \n",
       "NM_015665   8086.0        AAAS   \n",
       "NM_001087     14.0        AAMP   \n",
       "NM_025267  80755.0      AARSD1   \n",
       "NM_012089  23456.0      ABCB10   \n",
       "\n",
       "                                                   Gene name  Fold Change  \\\n",
       "NM_130786                             alpha-1-B glycoprotein     2.176788   \n",
       "NM_015665  achalasia, adrenocortical insufficiency, alacr...     1.952652   \n",
       "NM_001087           angio-associated, migratory cell protein     2.808783   \n",
       "NM_025267         alanyl-tRNA synthetase domain containing 1     1.962436   \n",
       "NM_012089  ATP-binding cassette, sub-family B (MDR/TAP), ...     0.225564   \n",
       "\n",
       "           q-value (%)  Negative Fold Change  \n",
       "NM_130786     1.998233                   NaN  \n",
       "NM_015665     3.753678                   NaN  \n",
       "NM_001087     0.637618                   NaN  \n",
       "NM_025267     9.865391                   NaN  \n",
       "NM_012089     1.707905             -4.433328  "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_148969</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.575574</td>\n",
       "      <td>6.349791</td>\n",
       "      <td>-1.737395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_496301</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.374709</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.668740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_181862</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.953817</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_145646</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.476765</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>-2.097469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_153032</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.087563</td>\n",
       "      <td>5.763981</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Geneid Gene symbol Gene name  Fold Change  q-value (%)  \\\n",
       "NM_148969     NaN         NaN       NaN     0.575574     6.349791   \n",
       "XM_496301     NaN         NaN       NaN     0.374709     3.399167   \n",
       "NM_181862     NaN         NaN       NaN     1.953817     8.270250   \n",
       "NM_145646     NaN         NaN       NaN     0.476765     8.270250   \n",
       "NM_153032     NaN         NaN       NaN     2.087563     5.763981   \n",
       "\n",
       "           Negative Fold Change  \n",
       "NM_148969             -1.737395  \n",
       "XM_496301             -2.668740  \n",
       "NM_181862                   NaN  \n",
       "NM_145646             -2.097469  \n",
       "NM_153032                   NaN  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Geneid', 'Gene symbol', 'Gene name', 'Fold Change', 'q-value (%)',\n",
       "       'Negative Fold Change'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#to get the column names\n",
    "genelist.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['NM_130786', 'NM_015665', 'NM_001087', 'NM_025267', 'NM_012089',\n",
       "       'NM_007189', 'NM_005692', 'NM_015407', 'NM_020676', 'NM_005759',\n",
       "       ...\n",
       "       'XM_372719', 'XM_498218', 'NM_016572', 'XM_379274', 'XM_114303',\n",
       "       'NM_148969', 'XM_496301', 'NM_181862', 'NM_145646', 'NM_153032'],\n",
       "      dtype='object', length=3085)"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# to get the row names\n",
    "genelist.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.0, 'A1BG', 'alpha-1-B glycoprotein', 2.17678774,\n",
       "        1.9982331000000002, nan],\n",
       "       [8086.0, 'AAAS',\n",
       "        'achalasia, adrenocortical insufficiency, alacrimia (Allgrove, triple-A)',\n",
       "        1.95265248, 3.753678, nan],\n",
       "       [14.0, 'AAMP', 'angio-associated, migratory cell protein',\n",
       "        2.80878298, 0.637618, nan],\n",
       "       ..., \n",
       "       [nan, nan, nan, 1.95381657, 8.27025, nan],\n",
       "       [nan, nan, nan, 0.47676513, 8.27025, -2.097468831],\n",
       "       [nan, nan, nan, 2.08756334, 5.7639809, nan]], dtype=object)"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3085, 6)"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2932.000000</td>\n",
       "      <td>3085.000000</td>\n",
       "      <td>3085.000000</td>\n",
       "      <td>838.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>43870.984993</td>\n",
       "      <td>1.764652</td>\n",
       "      <td>4.961999</td>\n",
       "      <td>-2.673922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>64141.163141</td>\n",
       "      <td>1.341177</td>\n",
       "      <td>3.038489</td>\n",
       "      <td>1.097826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.078751</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-12.698268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>5921.000000</td>\n",
       "      <td>0.542572</td>\n",
       "      <td>2.103816</td>\n",
       "      <td>-2.848319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>11316.000000</td>\n",
       "      <td>1.844670</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>-2.393723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>57175.000000</td>\n",
       "      <td>2.207867</td>\n",
       "      <td>8.039531</td>\n",
       "      <td>-2.094661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>654364.000000</td>\n",
       "      <td>24.535279</td>\n",
       "      <td>9.865391</td>\n",
       "      <td>-1.593398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Geneid  Fold Change  q-value (%)  Negative Fold Change\n",
       "count    2932.000000  3085.000000  3085.000000            838.000000\n",
       "mean    43870.984993     1.764652     4.961999             -2.673922\n",
       "std     64141.163141     1.341177     3.038489              1.097826\n",
       "min         1.000000     0.078751     0.000000            -12.698268\n",
       "25%      5921.000000     0.542572     2.103816             -2.848319\n",
       "50%     11316.000000     1.844670     4.611046             -2.393723\n",
       "75%     57175.000000     2.207867     8.039531             -2.094661\n",
       "max    654364.000000    24.535279     9.865391             -1.593398"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#good for numerical data, especially when you need to plot the data\n",
    "genelist.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that the summary table included summary statistics for Geneid which aren't really numerical values but ids for genes.  We will fix this shortly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indexing the Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we want to index the dataframe.  How do we get rows and columns that are interesting?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "Indexing is in this general format:\n",
    "dataframe[column name]\n",
    "dataframe.loc[rows, columns]\n",
    "\n",
    "#one column\n",
    "genelist['Geneid']\n",
    "\n",
    "#multiple columns\n",
    "genelist[['Geneid', 'Gene symbol']]\n",
    "\n",
    "#one row and one column    \n",
    "genelist.loc['NM_130786','Geneid']\n",
    "genelist.loc['NM_130786', 0]\n",
    "genelist.loc[0,0]\n",
    "\n",
    "#all values in one column\n",
    "genelist.loc[:,'Geneid']\n",
    "\n",
    "#all values in one row\n",
    "genelist.loc['NM_130786', :]\n",
    "\n",
    "#all values in multiple columns, put multiple items into list, must be consistent with datatypes in list\n",
    "genelist.loc[:, ['Geneid', 'Gene name']]\n",
    "genelist.loc[:, 0:3]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.loc['NM_130786','Geneid']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indexing exercise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Return the last column of the genelist dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NM_130786         NaN\n",
       "NM_015665         NaN\n",
       "NM_001087         NaN\n",
       "NM_025267         NaN\n",
       "NM_012089   -4.433328\n",
       "Name: Negative Fold Change, dtype: float64"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.iloc[:, -1].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Return NM_148969, NM_181862, XM_496301 rows of the genelist dataframe. Hint: this type of indexing will look similar in syntax to returning multiple columns. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_148969</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.575574</td>\n",
       "      <td>6.349791</td>\n",
       "      <td>-1.737395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_181862</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.953817</td>\n",
       "      <td>8.270250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_496301</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.374709</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.668740</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Geneid Gene symbol Gene name  Fold Change  q-value (%)  \\\n",
       "NM_148969     NaN         NaN       NaN     0.575574     6.349791   \n",
       "NM_181862     NaN         NaN       NaN     1.953817     8.270250   \n",
       "XM_496301     NaN         NaN       NaN     0.374709     3.399167   \n",
       "\n",
       "           Negative Fold Change  \n",
       "NM_148969             -1.737395  \n",
       "NM_181862                   NaN  \n",
       "XM_496301             -2.668740  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.loc[['NM_148969', 'NM_181862', 'XM_496301'], :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Return only the gene symbols from the previous question."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NM_148969    NaN\n",
       "NM_181862    NaN\n",
       "XM_496301    NaN\n",
       "Name: Gene symbol, dtype: object"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.loc[['NM_148969', 'NM_181862', 'XM_496301'], 'Gene symbol']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reassignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist['Geneid'].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NM_130786                      1.0\n",
       "NM_015665                   8086.0\n",
       "NM_001087                     14.0\n",
       "NM_025267                  80755.0\n",
       "NM_012089                  23456.0\n",
       "NM_007189                  10061.0\n",
       "NM_005692                  10061.0\n",
       "NM_015407                  25864.0\n",
       "NM_020676                  57406.0\n",
       "NM_005759                  10152.0\n",
       "NM_014945                  22885.0\n",
       "NM_013375                  29777.0\n",
       "NM_005891                     39.0\n",
       "NM_022914                  65057.0\n",
       "NM_181865                  11332.0\n",
       "NM_181866                  11332.0\n",
       "NM_181864                  11332.0\n",
       "NM_007274                  11332.0\n",
       "NM_004300                     52.0\n",
       "NM_177554                     52.0\n",
       "NM_007099                     52.0\n",
       "NM_203372                   2181.0\n",
       "NM_004457                   2181.0\n",
       "NM_005736                  10121.0\n",
       "NM_005735                  10120.0\n",
       "NM_020445                  57180.0\n",
       "NM_000666                     95.0\n",
       "NM_001107                     97.0\n",
       "NM_138448                     98.0\n",
       "NM_000022                    100.0\n",
       "                            ...   \n",
       "NM_021906                      nan\n",
       "NM_145323                      nan\n",
       "XM_039627                      nan\n",
       "NM_016496                  51257.0\n",
       "NM_148968                      nan\n",
       "XM_211529                      nan\n",
       "XM_372143                      nan\n",
       "NM_148973                      nan\n",
       "NM_148972                      nan\n",
       "NM_002097                      nan\n",
       "NM_015022                      nan\n",
       "XM_379250                      nan\n",
       "AFFX-HUMRGE/M10098_5_at        nan\n",
       "XM_374065                      nan\n",
       "XM_379694                      nan\n",
       "XM_371848                      nan\n",
       "XR_000223                      nan\n",
       "NR_002229                      nan\n",
       "XM_371838                      nan\n",
       "XM_499128                      nan\n",
       "XM_372719                      nan\n",
       "XM_498218                      nan\n",
       "NM_016572                      nan\n",
       "XM_379274                      nan\n",
       "XM_114303                      nan\n",
       "NM_148969                      nan\n",
       "XM_496301                      nan\n",
       "NM_181862                      nan\n",
       "NM_145646                      nan\n",
       "NM_153032                      nan\n",
       "Name: Geneid, Length: 3085, dtype: object"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist['Geneid'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist['Geneid'].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#reassinging values in a column in place\n",
    "genelist['Geneid'] = genelist['Geneid'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist['Geneid'].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>3085.000000</td>\n",
       "      <td>3085.000000</td>\n",
       "      <td>838.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.764652</td>\n",
       "      <td>4.961999</td>\n",
       "      <td>-2.673922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.341177</td>\n",
       "      <td>3.038489</td>\n",
       "      <td>1.097826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.078751</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-12.698268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.542572</td>\n",
       "      <td>2.103816</td>\n",
       "      <td>-2.848319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.844670</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>-2.393723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>2.207867</td>\n",
       "      <td>8.039531</td>\n",
       "      <td>-2.094661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>24.535279</td>\n",
       "      <td>9.865391</td>\n",
       "      <td>-1.593398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Fold Change  q-value (%)  Negative Fold Change\n",
       "count  3085.000000  3085.000000            838.000000\n",
       "mean      1.764652     4.961999             -2.673922\n",
       "std       1.341177     3.038489              1.097826\n",
       "min       0.078751     0.000000            -12.698268\n",
       "25%       0.542572     2.103816             -2.848319\n",
       "50%       1.844670     4.611046             -2.393723\n",
       "75%       2.207867     8.039531             -2.094661\n",
       "max      24.535279     9.865391             -1.593398"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genelist.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic plots of the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I would like to work only with the numerical data see the distribution of values for specific columns. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0,0.5,'Frequency')"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY4AAAEKCAYAAAAFJbKyAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAGDtJREFUeJzt3X+UX3V95/HnyyDyI9Yg6JQTsEGb\nWhFqlRG1uu4grfLDCrbQynJqcNF0z6K1q+dodN3Fbusp7oqoW3WNQg2uNaKosMIuRnRkOXtAiKLh\nh5YIKQQitAWiAyolvveP7x0ZxpnM907mO9+Z+T4f5+TMvZ/7uZ/7/uR7Mu/cz+d+PzdVhSRJ3Xpc\nvwOQJC0uJg5JUismDklSKyYOSVIrJg5JUismDklSKyYOSVIrPUscSS5Icm+SGyeU/bck30vy3SRf\nTLJiwrF3JNma5PtJXjGh/LimbGuSdb2KV5LUnV7ecXwSOG5S2SbgiKr6LeDvgXcAJDkceA3w7Oac\njyRZlmQZ8GHgeOBw4LSmriSpT/bqVcNVdVWSVZPKvjJh9xrglGb7JGBjVf0MuD3JVuDo5tjWqroN\nIMnGpu7Nu7v2QQcdVKtWrdpdlWk9+OCD7L///rM6dzGz34NlUPsNg9v3bvq9efPmf6qqp8zUVs8S\nRxf+LfDZZnslnUQybntTBnDnpPIXzNTwqlWruP7662cV1OjoKCMjI7M6dzGz34NlUPsNg9v3bvqd\n5B+6aasviSPJfwQeAT49XjRFtWLqobQpF9dKshZYCzA0NMTo6OisYhsbG5v1uYuZ/R4sg9pvGNy+\nz2W/5z1xJFkDvBI4th5dYXE7cOiEaocAdzfb05U/RlWtB9YDDA8P12z/R+H/RgaL/R48g9r3uez3\nvD6Om+Q44O3Aq6rqoQmHLgVek+QJSQ4DVgPfBK4DVic5LMnedCbQL53PmCVJj9WzO44knwFGgIOS\nbAfOpvMU1ROATUkArqmqf1dVNyW5iM6k9yPAWVW1q2nnjcAVwDLggqq6qVcxS5Jm1sunqk6bovj8\n3dR/D/CeKcovBy6fw9AkSXvAb45LkloxcUiSWjFxSJJaMXFIklrp5zfHl7xV6y6bsnzbOSfOcySS\nNHe845AktWLikCS1YuKQJLVi4pAktWLikCS1YuKQJLVi4pAktWLikCS1YuKQJLVi4pAktWLikCS1\nYuKQJLVi4pAktWLikCS1YuKQJLVi4pAktWLikCS1YuKQJLVi4pAktWLikCS1YuKQJLVi4pAktdKz\nxJHkgiT3JrlxQtmTk2xKcmvz84CmPEk+lGRrku8med6Ec9Y09W9NsqZX8UqSutPLO45PAsdNKlsH\nXFlVq4Erm32A44HVzZ+1wEehk2iAs4EXAEcDZ48nG0lSf/QscVTVVcB9k4pPAjY02xuAkyeUX1gd\n1wArkhwMvALYVFX3VdX9wCZ+ORlJkubRfM9xDFXVDoDm51Ob8pXAnRPqbW/KpiuXJPXJXv0OoJEp\nymo35b/cQLKWzjAXQ0NDjI6OziqQsbGxWZ872VuPfGTK8rlqfy7NZb8XE/s9eAa173PZ7/lOHPck\nObiqdjRDUfc25duBQyfUOwS4uykfmVQ+OlXDVbUeWA8wPDxcIyMjU1Wb0ejoKLM9d7Iz1l02Zfm2\n0+em/bk0l/1eTOz34BnUvs9lv+d7qOpSYPzJqDXAJRPKX9s8XfVCYGczlHUF8PIkBzST4i9vyiRJ\nfdKzO44kn6Fzt3BQku10no46B7goyZnAHcCpTfXLgROArcBDwOsAquq+JH8JXNfU+y9VNXnCXZI0\nj3qWOKrqtGkOHTtF3QLOmqadC4AL5jA0SdIe8JvjkqRWTBySpFZMHJKkVkwckqRWTBySpFZMHJKk\nVkwckqRWTBySpFZMHJKkVkwckqRWTBySpFZMHJKkVkwckqRWTBySpFZMHJKkVkwckqRWTBySpFZM\nHJKkVkwckqRWTBySpFZMHJKkVkwckqRWTBySpFZMHJKkVkwckqRWTBySpFZMHJKkVvqSOJL8hyQ3\nJbkxyWeS7JPksCTXJrk1yWeT7N3UfUKzv7U5vqofMUuSOuY9cSRZCfwZMFxVRwDLgNcA7wXOq6rV\nwP3Amc0pZwL3V9WvA+c19SRJfbJXH6+7b5J/AfYDdgAvA/5Nc3wD8G7go8BJzTbA54G/SZKqql4F\nt2rdZVOWbzvnxF5dUpIWjXm/46iqu4D3AXfQSRg7gc3AA1X1SFNtO7Cy2V4J3Nmc+0hT/8D5jFmS\n9Kj08D/uU18wOQC4GPhj4AHgc83+2c1wFEkOBS6vqiOT3AS8oqq2N8d+ABxdVf88qd21wFqAoaGh\nozZu3Dir+MbGxrh9564pjx258kmt2tpy1845aWc+jI2NsXz58n6HMe/s9+AZ1L530+9jjjlmc1UN\nz9RWP4aqfhe4var+ESDJF4DfAVYk2au5qzgEuLupvx04FNieZC/gScB9kxutqvXAeoDh4eEaGRmZ\nVXCjo6Oce/WDUx7bdnq7Ns+YbsirZTvzYXR0lNn+nS1m9nvwDGrf57Lf/Xiq6g7ghUn2SxLgWOBm\n4OvAKU2dNcAlzfalzT7N8a/1cn5DkrR7/ZjjuJbOJPe3gC1NDOuBtwNvSbKVzhzG+c0p5wMHNuVv\nAdbNd8ySpEf15amqqjobOHtS8W3A0VPU/Slw6nzEJUmamd8clyS1YuKQJLVi4pAktWLikCS1YuKQ\nJLVi4pAktdJV4khyRK8DkSQtDt3ecfyPJN9M8u+TrOhpRJKkBa2rxFFVLwFOp7Nm1PVJ/i7J7/U0\nMknSgtT1HEdV3Qq8i87SIP8a+FCS7yX5g14FJ0laeLqd4/itJOcBt9B54dLvV9Wzmu3zehifJGmB\n6Xatqr8BPg68s6p+Ml5YVXcneVdPIpMkLUjdJo4TgJ9U1S6AJI8D9qmqh6rqUz2LTpK04HQ7x/FV\nYN8J+/s1ZZKkAdNt4tinqsbGd5rt/XoTkiRpIes2cTyY5HnjO0mOAn6ym/qSpCWq2zmOPwc+l2T8\nPeAHA3/cm5AkSQtZV4mjqq5L8pvAM4EA36uqf+lpZJKkBanNq2OfD6xqznluEqrqwp5EJUlasLpK\nHEk+BTwDuAHY1RQXYOKQpAHT7R3HMHB4VVUvg5EkLXzdPlV1I/CrvQxEkrQ4dHvHcRBwc5JvAj8b\nL6yqV/UkKknSgtVt4nh3L4OQJC0e3T6O+40kvwasrqqvJtkPWNbb0CRJC1G3y6q/Afg88LGmaCXw\npV4FJUlauLqdHD8LeDHwI/jFS52e2qugJEkLV7eJ42dV9fD4TpK96HyPY1aSrEjy+eYNgrckeVGS\nJyfZlOTW5ucBTd0k+VCSrUm+O3HNLEnS/Os2cXwjyTuBfZt3jX8O+F97cN0PAv+nqn4TeA6dNwuu\nA66sqtXAlc0+wPHA6ubPWuCje3BdSdIe6jZxrAP+EdgC/ClwOZ33j7eW5FeAlwLnA1TVw1X1AHAS\nsKGptgE4udk+CbiwOq4BViQ5eDbXliTtuW6fqvo5nVfHfnwOrvl0Oknob5M8B9gMvBkYqqodzfV2\nJBmfQ1kJ3Dnh/O1N2Y45iEWS1FK6WUUkye1MMadRVU9vfcFkGLgGeHFVXZvkg3Qm3d9UVSsm1Lu/\nqg5Ichnw11V1dVN+JfC2qto8qd21dIayGBoaOmrjxo1tQwNgbGyM23fumvLYkSuf1KqtLXftnJN2\n5sPY2BjLly/vdxjzzn4PnkHtezf9PuaYYzZX1fBMbbVZq2rcPsCpwJO7PHey7cD2qrq22f88naGw\ne5Ic3NxtHAzcO6H+oRPOPwS4m0mqaj2wHmB4eLhGRkZmFdzo6CjnXv3glMe2nd6uzTPWXTb1gS3T\ntH/Oia3an0ujo6PM9u9sMbPfg2dQ+z6X/e5qjqOq/nnCn7uq6gPAy2Zzwar6IXBnkmc2RccCNwOX\nAmuasjXAJc32pcBrm6erXgjsHB/SkiTNv26XVZ/4COzj6NyBPHEPrvsm4NNJ9gZuA17XtHtRkjOB\nO+jc1UBnIv4EYCvwUFNXktQn3Q5VnTth+xFgG/BHs71oVd3AY4e/xh07Rd2i8wVESdIC0O1TVcf0\nOhBJ0uLQ7VDVW3Z3vKrePzfhSJIWujZPVT2fzkQ1wO8DV/HY71dIkgZAmxc5Pa+qfgyQ5N3A56rq\n9b0KTJK0MHW75MjTgIcn7D8MrJrzaCRJC163dxyfAr6Z5It0vkH+auDCnkUlSVqwun2q6j1J/jfw\nr5qi11XVt3sXliRpoep2qApgP+BHVfVBYHuSw3oUkyRpAev21bFnA28H3tEUPR74n70KSpK0cHV7\nx/Fq4FXAgwBVdTd7tuSIJGmR6jZxPNws/VEASfbvXUiSpIWs28RxUZKP0Xn73huArzI3L3WSJC0y\n3T5V9b7mXeM/Ap4J/Oeq2tTTyCRJC9KMiSPJMuCKqvpdwGQhSQNuxqGqqtoFPJRk4b3vVJI077r9\n5vhPgS1JNtE8WQVQVX/Wk6gkSQtWt4njsuaPJGnA7TZxJHlaVd1RVRvmKyBJ0sI20xzHl8Y3klzc\n41gkSYvATIkjE7af3stAJEmLw0yJo6bZliQNqJkmx5+T5Ed07jz2bbZp9quqfqWn0UmSFpzdJo6q\nWjZfgUiSFoc27+OQJMnEIUlqx8QhSWrFxCFJaqVviSPJsiTfTvLlZv+wJNcmuTXJZ5Ps3ZQ/odnf\n2hxf1a+YJUn9veN4M3DLhP33AudV1WrgfuDMpvxM4P6q+nXgvKaeJKlP+pI4khwCnAh8otkP8DLg\n802VDcDJzfZJzT7N8WOb+pKkPujXHccHgLcBP2/2DwQeqKpHmv3twMpmeyVwJ0BzfGdTX5LUB90u\nqz5nkrwSuLeqNicZGS+eomp1cWxiu2uBtQBDQ0OMjo7OKr6xsTHeeuSuKY+1bfOtRz4yc6U9aH8u\njY2N9fX6/WK/B8+g9n0u+z3viQN4MfCqJCcA+wC/QucOZEWSvZq7ikOAu5v624FDge1J9gKeBNw3\nudGqWg+sBxgeHq6RkZFZBTc6Osq5Vz845bFtp7dr84x17V5h0rb9uTQ6Osps/84WM/s9eAa173PZ\n73kfqqqqd1TVIVW1CngN8LWqOh34OnBKU20NcEmzfWmzT3P8a1XlgouS1CcL6XscbwfekmQrnTmM\n85vy84EDm/K3AOv6FJ8kif4MVf1CVY0Co832bcDRU9T5KXDqvAYmSZrWQrrjkCQtAiYOSVIrJg5J\nUismDklSKyYOSVIrJg5JUismDklSKyYOSVIrJg5JUismDklSKyYOSVIrJg5JUismDklSKyYOSVIr\nJg5JUismDklSKyYOSVIrJg5JUismDklSKyYOSVIrJg5JUismDklSKyYOSVIrJg5JUismDklSK3v1\nOwA9atW6y6Ys33bOifMciSRNzzsOSVIr8544khya5OtJbklyU5I3N+VPTrIpya3NzwOa8iT5UJKt\nSb6b5HnzHbMk6VH9uON4BHhrVT0LeCFwVpLDgXXAlVW1Griy2Qc4Hljd/FkLfHT+Q5YkjZv3xFFV\nO6rqW832j4FbgJXAScCGptoG4ORm+yTgwuq4BliR5OB5DluS1OjrHEeSVcBzgWuBoaraAZ3kAjy1\nqbYSuHPCadubMklSH6Sq+nPhZDnwDeA9VfWFJA9U1YoJx++vqgOSXAb8dVVd3ZRfCbytqjZPam8t\nnaEshoaGjtq4ceOs4hobG+P2nbumPHbkyie1amvLXTtnFcOeXnc2xsbGWL58ec+vs9DY78EzqH3v\npt/HHHPM5qoanqmtvjyOm+TxwMXAp6vqC03xPUkOrqodzVDUvU35duDQCacfAtw9uc2qWg+sBxge\nHq6RkZFZxTY6Osq5Vz845bFtp7dr84xpHq9tq+11Z2N0dJTZ/p0tZvZ78Axq3+ey3/14qirA+cAt\nVfX+CYcuBdY022uASyaUv7Z5uuqFwM7xIS1J0vzrxx3Hi4E/AbYkuaEpeydwDnBRkjOBO4BTm2OX\nAycAW4GHgNfNb7iSpInmPXE0cxWZ5vCxU9Qv4KyeBiVJ6prfHJcktWLikCS1YuKQJLVi4pAktWLi\nkCS1YuKQJLVi4pAktWLikCS1YuKQJLVi4pAktWLikCS1YuKQJLVi4pAktWLikCS1YuKQJLXSl1fH\nqp1V07yCdts5J85zJJLkHYckqSUThySpFROHJKkV5zgWMec+JPWDdxySpFa845gD0/3PX5KWIu84\nJEmtmDgkSa2YOCRJrZg4JEmtODnegpPgkrSIEkeS44APAsuAT1TVOX0OacHy+x2SemlRDFUlWQZ8\nGDgeOBw4Lcnh/Y1KkgbTYrnjOBrYWlW3ASTZCJwE3NzXqBaZ3Q21ffK4/ecxEkmL2WJJHCuBOyfs\nbwde0KdYlqQtd+3kjDmYw1ksw2HjSfStRz7yS/1eLH3Q7Dmcu2dSVf2OYUZJTgVeUVWvb/b/BDi6\nqt40oc5aYG2z+0zg+7O83EHAP+1BuIuV/R4sg9pvGNy+d9PvX6uqp8zU0GK549gOHDph/xDg7okV\nqmo9sH5PL5Tk+qoa3tN2Fhv7PVgGtd8wuH2fy34vislx4DpgdZLDkuwNvAa4tM8xSdJAWhR3HFX1\nSJI3AlfQeRz3gqq6qc9hSdJAWhSJA6CqLgcun4dL7fFw1yJlvwfLoPYbBrfvc9bvRTE5LklaOBbL\nHIckaYEwcTSSHJfk+0m2JlnX73jmU5JtSbYkuSHJ9f2Op1eSXJDk3iQ3Tih7cpJNSW5tfh7Qzxh7\nYZp+vzvJXc1nfkOSE/oZYy8kOTTJ15PckuSmJG9uypf0Z76bfs/ZZ+5QFb9Y0uTvgd+j8+jvdcBp\nVTUQ30xPsg0Yrqol/Wx7kpcCY8CFVXVEU/Zfgfuq6pzmPwwHVNXb+xnnXJum3+8Gxqrqff2MrZeS\nHAwcXFXfSvJEYDNwMnAGS/gz302//4g5+sy94+j4xZImVfUwML6kiZaQqroKuG9S8UnAhmZ7A51/\nYEvKNP1e8qpqR1V9q9n+MXALnVUolvRnvpt+zxkTR8dUS5rM6V/0AlfAV5Jsbr6BP0iGqmoHdP7B\nAU/tczzz6Y1JvtsMZS2p4ZrJkqwCngtcywB95pP6DXP0mZs4OjJF2SCN4b24qp5HZ/Xhs5qhDS1t\nHwWeAfw2sAM4t7/h9E6S5cDFwJ9X1Y/6Hc98maLfc/aZmzg6ZlzSZCmrqrubn/cCX6QzdDco7mnG\nhMfHhu/tczzzoqruqapdVfVz4OMs0c88yePp/PL8dFV9oSle8p/5VP2ey8/cxNExsEuaJNm/mUAj\nyf7Ay4Ebd3/WknIpsKbZXgNc0sdY5s34L87Gq1mCn3mSAOcDt1TV+yccWtKf+XT9nsvP3KeqGs2j\naR/g0SVN3tPnkOZFkqfTucuAzkoCf7dU+57kM8AInVVC7wHOBr4EXAQ8DbgDOLWqltRE8jT9HqEz\nZFHANuBPx8f9l4okLwH+L7AF+HlT/E464/1L9jPfTb9PY44+cxOHJKkVh6okSa2YOCRJrZg4JEmt\nmDgkSa2YOCRJrZg4NLCS7JqwUugNzfIM09UdSfLlaY5tS3LQFOXLk3wsyQ+aVUqvSvKCJKsmrlQr\nLTaL5g2AUg/8pKp+u4ftfwK4HVhdVT9vvjPzLDrfpZAWLe84pAmS7JPkb5v3k3w7yTFT1DkwyVea\n4x9jirXOkjwDeAHwrmaJB5rVly9rqixL8vHmTuQrSfZtzntDkuuSfCfJxUn2a8o/meRDSf5fktuS\nnNKUPy7JR5p2vpzk8gnHjkryjWbxyismfXNYmjUThwbZvhOGqca/PX8WQFUdSeebthuS7DPpvLOB\nq6vquXSWr3jaFG0/G7ihqnZNc+3VwIer6tnAA8AfNuVfqKrnV9Vz6CyHfeaEcw4GXgK8EjinKfsD\nYBVwJPB64EXwi7WK/jtwSlUdBVwALMkVATT/HKrSIJtqqOoldH7hUlXfS/IPwG9MqvNSOr+wqarL\nktw/i2vfXlU3NNub6fzyBzgiyV8BK4DlwBUTzvlSc/dyc5KhCfF+rin/YZKvN+XPBI4ANnWWLmIZ\nnRVRpT1m4pAea6ol9qcy01o9NwHPSfK48aGqSX42YXsXsG+z/Ung5Kr6TpIz6KwpNdU5mfRzsgA3\nVdWLZohTas2hKumxrgJOB0jyG3SGob6/mzrHA7/0Qpyq+gFwPfAXzWqlJFmdZKY3Sz4R2NEMNZ3e\nRbxXA3/YzHUM8Wii+T7wlCS/GLpK8uwu2pNmZOKQHusjdCautwCfBc6oqp9NqvMXwEuTfIvOMvR3\nTNPW64FfBbY27X2cmd/z8p/orN66CfheF/FeTOd9MjcCH2vO3dm8AvkU4L1JvgPcAPxOF+1JM3J1\nXGmRS7K8qsaSHAh8k84bHX/Y77i0dDnHIS1+X06yAtgb+EuThnrNOw5JUivOcUiSWjFxSJJaMXFI\nkloxcUiSWjFxSJJaMXFIklr5/7ut4XVaDSAuAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10fe94b38>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ax = genelist['Fold Change'].hist(bins=50)\n",
    "ax.set_xlabel('Fold Change')\n",
    "ax.set_ylabel('Frequency')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that there are a lot of values around zero and no negative values.  Meaning if we want to at some point look at negative fold-changes the authors have wierdly put those in the adjacent column.  Also, the authors titled this figure, \"Significant translational changes\", without defining significant.  So, lets try to see if they used some sort of FDR %threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0,0.5,'Frequency')"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEKCAYAAAAIO8L1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAFydJREFUeJzt3XuQpXV95/H3RyAKjAEstBcHdHR3\nQkTGILSWG2N2RrLeFUkFg0UUvE22lhjdsKujm1KqXDZsyktiTExQjKjISPACCsbgrCNrlYozyDrg\n6EoJgRkm4IUMDBJx8Lt/nKfjoXm6+3TPefp0z3m/qrr6Ob9+Lt/fnDn96ef2e1JVSJI03cNGXYAk\naWkyICRJrQwISVIrA0KS1MqAkCS1MiAkSa0MCElSKwNCktTKgJAktTpw1AXsiyOPPLJWrVq1oGXv\nvfdeDj300OEWtAzY7/Eyjv0exz7D/Pq9devWH1bVo+eab1kHxKpVq9iyZcuClt28eTNr164dbkHL\ngP0eL+PY73HsM8yv30n+cZD5PMQkSWplQEiSWhkQkqRWBoQkqZUBIUlqZUBIkloZEJKkVp0FRJJj\nknwpyfYkNyZ5Q9N+bpKdSa5vvl7Qt8xbktyU5LtJnttVbZKkuXV5o9xe4Jyqui7JI4GtSa5ufvae\nqnpn/8xJjgNOB54MPBb4YpJfqaoHOqxRkjSDzgKiqnYBu5rpe5JsB1bOssgpwMaq+ilwc5KbgKcD\nX+2qxmFZteHK1vZbzn/hIlciScOTqup+I8kq4BrgeOCPgLOAu4Et9PYy7kryPuBrVfWxZpkLgc9X\n1WXT1rUeWA8wMTFx0saNGxdU0549e1ixYsWClp1u287dre1rVh42lPUP0zD7vZzY7/Exjn2G+fV7\n3bp1W6tqcq75Oh+LKckK4JPAG6vq7iTvB94BVPP9XcCrgbQs/pD0qqoLgAsAJicna6FjrgxzvJaz\nZtqDOGM46x8mx6kZL+PY73HsM3TT706vYkpyEL1wuLiqPgVQVXdU1QNV9XPgA/QOIwHsAI7pW/xo\n4PYu65MkzazLq5gCXAhsr6p397Uf1TfbqcANzfQVwOlJHp7kCcBq4Nqu6pMkza7LQ0zPBF4BbEty\nfdP2VuDlSU6gd/joFuD3AarqxiSXAt+mdwXU2V7BJEmj0+VVTF+h/bzCVbMscx5wXlc1SZIG553U\nkqRWBoQkqZUBIUlqZUBIkloZEJKkVgaEJKmVASFJamVASJJaGRCSpFYGhCSplQEhSWplQEiSWhkQ\nkqRWBoQkqZUBIUlqZUBIkloZEJKkVgaEJKmVASFJamVASJJaGRCSpFYGhCSplQEhSWplQEiSWhkQ\nkqRWBoQkqZUBIUlqZUBIkloZEJKkVgaEJKmVASFJamVASJJadRYQSY5J8qUk25PcmOQNTfujklyd\n5HvN9yOa9iR5b5KbknwryYld1SZJmluXexB7gXOq6knAM4CzkxwHbAA2VdVqYFPzGuD5wOrmaz3w\n/g5rkyTNobOAqKpdVXVdM30PsB1YCZwCXNTMdhHw0mb6FOAj1fM14PAkR3VVnyRpdqmq7jeSrAKu\nAY4Hbq2qw/t+dldVHZHkc8D5VfWVpn0T8Oaq2jJtXevp7WEwMTFx0saNGxdU0549e1ixYsWClp1u\n287dre1rVh42lPUP0zD7vZzY7/Exjn2G+fV73bp1W6tqcq75DtznquaQZAXwSeCNVXV3khlnbWl7\nSHpV1QXABQCTk5O1du3aBdW1efNmFrrsdGdtuLK1/ZYzhrP+YRpmv5cT+z0+xrHP0E2/O72KKclB\n9MLh4qr6VNN8x9Sho+b7nU37DuCYvsWPBm7vsj5J0sy6vIopwIXA9qp6d9+PrgDObKbPBC7va39l\nczXTM4DdVbWrq/okSbPr8hDTM4FXANuSXN+0vRU4H7g0yWuAW4HTmp9dBbwAuAn4CfCqDmuTJM2h\ns4BoTjbPdMLh5Jb5Czi7q3okSfPjndSSpFYGhCSplQEhSWplQEiSWnV+o5wkjcqqmW5iPf+Fi1zJ\n8uQehCSplQEhSWplQEiSWhkQkqRWBoQkqZUBIUlqZUBIkloZEJKkVgaEJKmVASFJamVASJJaGRCS\npFYGhCSplQEhSWplQEiSWhkQkqRWBoQkqZUBIUlq5SNHJWkO4/roUvcgJEmtDAhJUquBAiLJ8V0X\nIklaWgbdg/jrJNcm+c9JDu+0IknSkjBQQFTVbwBnAMcAW5J8PMl/7LQySdJIDXwOoqq+B/wx8Gbg\nPwDvTfKdJL/dVXGSpNEZ9BzEU5K8B9gOPBt4cVU9qZl+T4f1SZJGZND7IN4HfAB4a1XdN9VYVbcn\n+eNOKpMkjdSgh5heAHx8KhySPCzJIQBV9dG2BZJ8KMmdSW7oazs3yc4k1zdfL+j72VuS3JTku0me\nu/AuSZKGYdCA+CJwcN/rQ5q22XwYeF5L+3uq6oTm6yqAJMcBpwNPbpb5qyQHDFibJKkDgwbEI6pq\nz9SLZvqQ2RaoqmuAHw+4/lOAjVX106q6GbgJePqAy0qSOjBoQNyb5MSpF0lOAu6bZf7Z/EGSbzWH\noI5o2lYCt/XNs6NpkySNSKpq7pmSpwEbgdubpqOA362qrXMstwr4XFUd37yeAH4IFPAO4KiqenWS\nvwS+WlUfa+a7ELiqqj7Zss71wHqAiYmJkzZu3DhANx9qz549rFixYkHLTrdt5+7W9jUrDxvK+odp\nmP1eTuz3+Ojv87A+m8vhMz6f93rdunVbq2pyrvkGuoqpqr6R5FeBY4EA36mqnw1UyYPXc8fUdJIP\nAJ9rXu6gdxPelKP5RRhNX8cFwAUAk5OTtXbt2vmWAcDmzZtZ6LLTnTXTSI9nDGf9wzTMfi8n9nt8\n9Pd5WJ/N5fAZ7+K9ns9gfU8DngI8FXh5klfOd2NJjup7eSowdYXTFcDpSR6e5AnAauDa+a5fkjQ8\nA+1BJPko8G+B64EHmuYCPjLLMpcAa4Ejk+wA3g6sTXJCs+wtwO8DVNWNSS4Fvg3sBc6uqgfa1itJ\nWhyD3ig3CRxXg5ywaFTVy1uaL5xl/vOA8wZdvySpW4MeYroB+DddFiJJWloG3YM4Evh2kmuBn041\nVtVLOqlKAxnXxyBKWhyDBsS5XRYhSVp6Br3M9ctJHg+srqovNuMwORSGJO3HBh3u+3XAZcDfNE0r\ngc90VZQkafQGPcR0Nr2xkb4OvYcHJXlMZ1VJ0gjMdF5vXA16FdNPq+r+qRdJDqR3L4MkaT81aEB8\nOclbgYObZ1H/HfDZ7sqSJI3aoAGxAfgBsI3e3c9X0Xs+tSRpPzXoVUw/p/fI0Q90W44kaakYdCym\nm2k551BVTxx6RZKkJWE+YzFNeQRwGvCo4ZcjSVoqBjoHUVU/6vvaWVV/Bjy749okSSM06CGmE/te\nPozeHsUjO6lIkrQkDHqI6V1903vpPcvhZUOvRpK0ZAx6FdO6rgtR91ZtuJJz1ux9yOMTHf1VUptB\nDzH90Ww/r6p3D6ccaWEc+lwavvlcxfQ0es+OBngxcA1wWxdFSZJGbz4PDDqxqu4BSHIu8HdV9dqu\nCpMkjdagQ208Dri/7/X9wKqhVyNJWjIG3YP4KHBtkk/Tu6P6VOAjnVUlSRq5Qa9iOi/J54FnNU2v\nqqpvdleWJGnUBt2DADgEuLuq/jbJo5M8oapu7qowSbPzyi11bdBHjr4deDPwlqbpIOBjXRUlSRq9\nQU9Snwq8BLgXoKpux6E2JGm/NmhA3F9VRTPkd5JDuytJkrQUDBoQlyb5G+DwJK8DvogPD5Kk/dqg\nVzG9s3kW9d3AscDbqurqTiuTJI3UnAGR5ADgC1X1W4ChIEljYs5DTFX1APCTJIctQj2SpCVi0Psg\n/gXYluRqmiuZAKrqDzupSpI0coMGxJXNlyRpTMwaEEkeV1W3VtVF811xkg8BLwLurKrjm7ZHAZ+g\nN9DfLcDLququJAH+HHgB8BPgrKq6br7blLR/8C7xpWGucxCfmZpI8sl5rvvDwPOmtW0ANlXVamBT\n8xrg+cDq5ms98P55bkuSNGRzBUT6pp84nxVX1TXAj6c1nwJM7Y1cBLy0r/0j1fM1evdbHDWf7UmS\nhmuugKgZphdqoqp2ATTfH9O0r+TBT6fb0bRJkkYkvRE0Zvhh8gC9q5YCHEzv/ADN66qqX5515ckq\n4HN95yD+uaoO7/v5XVV1RJIrgT+pqq807ZuAN1XV1pZ1rqd3GIqJiYmTNm7cOGBXH2zPnj2sWLFi\nQctOt23n7tb2NSu7vTJ4vtvdtnM3EwfDHfcNNv9yMte/xTDf76VikPd/ufZ7Xz5T/X1eyGdkPpbS\nZ2c+7/W6deu2VtXkXPPNepK6qg4YsLZB3ZHkqKra1RxCurNp3wEc0zff0cDtM9R0AXABwOTkZK1d\nu3ZBhWzevJmFLjvdWTOdUDtjOOsf1nbP2nAl56zZy7u2HTjQ/MvJXP8Ww3y/l4pB3v/l2u99+Uz1\n93khn5H5WEqfnS7e60HHYhqWK4Azm+kzgcv72l+ZnmcAu6cORUmSRmM+DwyalySXAGuBI5PsAN4O\nnE9v4L/XALcCpzWzX0XvEteb6B3GelVXdUmSBtNZQFTVy2f40ckt8xZwdle1SJLmr7OAkJaCqRuu\nzlmz90HHlxfjhitv9tJyt9jnICRJy4QBIUlqZUBIkloZEJKkVp6klsacJ9M1E/cgJEmt3IPQjGb6\ny3I2/tUpDdco9/AMCEnLXv8v0en3vGjhPMQkSWplQEiSWhkQkqRWnoNYBhZysliS9pV7EJKkVu5B\nSNKQzbbXv5wuBTcgJHXOu7WXJw8xSZJaGRCSpFYGhCSplQEhSWrlSeolxPsdJC0l7kFIkloZEJKk\nVgaEJKmVASFJamVASJJaeRWTJC0BS/EqRgNC0rw4rtL4MCC0JHX9S2i+6x/l6JxL8S9LjQfPQUiS\nWrkHoZHyr2Np6TIgJGkRLac/ikYSEEluAe4BHgD2VtVkkkcBnwBWAbcAL6uqu0ZRnyRptOcg1lXV\nCVU12bzeAGyqqtXApua1JGlEltJJ6lOAi5rpi4CXjrAWSRp7owqIAv4hydYk65u2iaraBdB8f8yI\napMkAamqxd9o8tiquj3JY4CrgdcDV1TV4X3z3FVVR7Qsux5YDzAxMXHSxo0bF1TDnT/ezR33tf9s\nzcrD5rWubTt3d7qe+Zppu9t27mbiYB7S79nmH9a2ZzLfbSy01rZ+D2v9811mMf+N9uzZw4oVKxa0\n/mH2q+tt9M8/yHs9rP/zw/zszNf0bQ/yXk9Zt27d1r7D+zMaSUA8qIDkXGAP8DpgbVXtSnIUsLmq\njp1t2cnJydqyZcuCtvsXF1/Ou7a1n6Of741Pw7qpa1hXN8x2s9c5a/Y+pN8LuTlsvtueyXy3sdBa\n2/o9rPXPd5nF/DfavHkza9euXdD6h9mvrrfRP/8g7/Ww/s8P87MzX9O3Pch7PSXJQAGx6IeYkhya\n5JFT08BzgBuAK4Azm9nOBC5f7NokSb8wistcJ4BPJ5na/ser6u+TfAO4NMlrgFuB00ZQmySpsegB\nUVXfB36tpf1HwMmLXY8Wx3K6OUhSj3dSa1kxaKTFs5Tug5AkLSEGhCSplQEhSWrlOQhJrTzfIwNi\nP+QHe3nyfdNS4yEmSVIrA0KS1MpDTPLQhqRW7kFIklq5B9GhYY3mKUmj4B6EJKmVexAj4DF/ScuB\nexCSpFYGhCSplQEhSWplQEiSWhkQkqRWBoQkqZUBIUlqZUBIkloZEJKkVgaEJKmVASFJamVASJJa\nGRCSpFYGhCSplQEhSWplQEiSWhkQkqRWBoQkqZUBIUlqteQCIsnzknw3yU1JNoy6HkkaV0sqIJIc\nAPwl8HzgOODlSY4bbVWSNJ6WVEAATwduqqrvV9X9wEbglBHXJEljaakFxErgtr7XO5o2SdIiS1WN\nuoZ/leQ04LlV9drm9SuAp1fV6/vmWQ+sb14eC3x3gZs7EvjhPpS7XNnv8TKO/R7HPsP8+v34qnr0\nXDMduG/1DN0O4Ji+10cDt/fPUFUXABfs64aSbKmqyX1dz3Jjv8fLOPZ7HPsM3fR7qR1i+gawOskT\nkvwScDpwxYhrkqSxtKT2IKpqb5I/AL4AHAB8qKpuHHFZkjSWllRAAFTVVcBVi7CpfT5MtUzZ7/Ey\njv0exz5DB/1eUiepJUlLx1I7ByFJWiLGMiDGcTiPJMck+VKS7UluTPKGUde0WJIckOSbST436loW\nS5LDk1yW5DvNe/7vR13TYkjyX5r/3zckuSTJI0ZdUxeSfCjJnUlu6Gt7VJKrk3yv+X7Evm5n7AJi\njIfz2AucU1VPAp4BnD0m/QZ4A7B91EUssj8H/r6qfhX4Ncag/0lWAn8ITFbV8fQudDl9tFV15sPA\n86a1bQA2VdVqYFPzep+MXUAwpsN5VNWuqrqumb6H3i+M/f4u9SRHAy8EPjjqWhZLkl8GfhO4EKCq\n7q+qfx5tVYvmQODgJAcChzDtPqr9RVVdA/x4WvMpwEXN9EXAS/d1O+MYEGM/nEeSVcBTga+PtpJF\n8WfAm4Cfj7qQRfRE4AfA3zaH1j6Y5NBRF9W1qtoJvBO4FdgF7K6qfxhtVYtqoqp2Qe8PQuAx+7rC\ncQyItLSNzaVcSVYAnwTeWFV3j7qeLiV5EXBnVW0ddS2L7EDgROD9VfVU4F6GcLhhqWuOuZ8CPAF4\nLHBokt8bbVXL2zgGxJzDeeyvkhxELxwurqpPjbqeRfBM4CVJbqF3KPHZST422pIWxQ5gR1VN7SFe\nRi8w9ne/BdxcVT+oqp8BnwJ+fcQ1LaY7khwF0Hy/c19XOI4BMZbDeSQJvWPS26vq3aOuZzFU1Vuq\n6uiqWkXvff7fVbXf/0VZVf8E3Jbk2KbpZODbIyxpsdwKPCPJIc3/95MZg5Pzfa4AzmymzwQu39cV\nLrk7qbs2xsN5PBN4BbAtyfVN21ubO9e1/3k9cHHzR9D3gVeNuJ7OVdXXk1wGXEfvqr1vsp/eVZ3k\nEmAtcGSSHcDbgfOBS5O8hl5YnrbP2/FOaklSm3E8xCRJGoABIUlqZUBIkloZEJKkVgaEJKmVASE1\nkmxO8txpbW9M8ldzLLdniDW8Mckrm+n/leRbST7S9/NX9I/Em2RNkg8Pa/tSPwNC+oVLeOjon6c3\n7Z1rBph7NfDxJIcBv15VTwEOaILgYOAs4F8Dq6q2AUcnedxi1KjxYkBoLCT5780zQL7YPCfgv7bM\ndhnwoiQPb5ZZRW9Mn68kWZFkU5LrkmxL8pARgJOs7X/mRJL3JTmrmT4pyZeTbE3yhakhEaZ5NnBd\nVe2lN7jgLzV3BB8M/Az4b8B7m2Ek+n2W/XdYa42QAaH9XpKT6P0CfSrw28DT2uarqh8B1/KLcfZP\nBz5RvbtJ/wU4tapOBNYB72p+eQ+y/YOAvwB+p6pOAj4EnNcy6zOBrU0t99AbN+ubwM3AbuBpVdU2\nfMIW4FmD1CLNx9gNtaGx9Czg01X1E4Aks429NXWY6fLm+6ub9gD/M8lv0vvrfiUwAfzTANs/Fjge\nuLrJlAPoDUc93VH0jR1UVX8K/GlT8weBtyV5LfAc4FtV9T+aWe+kt6cjDZV7EBoXDxlTpnkM6/XN\n139qmj8DnJzkRODgqYcsAWcAjwZOqqoTgDuA6Y+z3MuDP1NTPw9wY1Wd0HytqarntNR4X8s6SfLU\nZvL/Aa+sqpcBxydZ3bed+2buurQwBoTGwTXAqUkOTvJI4MUAVXVb3y/tv27a9gCb6R0G6j85fRi9\nZ0v8LMk64PEt2/lH4LgkD29OMp/ctH8XePTUc6GTHJTkyS3Lbwf+XUv7O4C3AQfR2/uA3l7MIc30\nrwA3tCwn7RMPMWm/V1XXJfkEcD29X+L/Z45FLqH3LIH+E78XA59NsqVZz3datnNbkkuBbwHfo3f+\ngKq6P8nvAO9tguNAek+6mz6K8OeBj/Y3JHkp8I2qur15/dUk2+gdYvq/zWzrgCvn6JM0b47mqrGT\n5FxgT1W9c9S1TJfk08Cbqup7A87/cODLwG80Vz9JQ+MhJmlp2UDvZPWgHgdsMBzUBfcgJEmt3IOQ\nJLUyICRJrQwISVIrA0KS1MqAkCS1MiAkSa3+PxQrb68izLFYAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10fea0710>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ax2 = genelist['q-value (%)'].hist(bins=50)\n",
    "ax2.set_xlabel('q-Value (%)')\n",
    "ax2.set_ylabel('Frequency')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see now that there are some weird things in the data.  There are genes that lack names, which are not useful for downstream pathway analysis and for reasons unknown, the authors provided downregulated genes as negative values or fractions of one.  So we are going to try and fix some of these things.  We also have 3,085 probe values to work with, and if you've ever tried to validate 3000 target genes at the bench, I'm feeling bad for you.  So we are going to try and reduce our targets trying several strategies. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering the data for smaller gene-list using conditionals in Pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some ways to reduce the gene-list:  \n",
    "1. Filter out genes without gene symbols\n",
    "2. Make the q-value cut-off more strict\n",
    "3. Employ a fold-change cut-off"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by q-value"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pull out rows where q-value is < 5% by using a Boolean mask.  Here is what the mask looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NM_130786     True\n",
       "NM_015665     True\n",
       "NM_001087     True\n",
       "NM_025267    False\n",
       "NM_012089     True\n",
       "Name: q-value (%), dtype: bool"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask = genelist['q-value (%)']<5\n",
    "mask.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To index into the dataframe, plug the mask into the rows index:\n",
    "```python\n",
    "dataframe.loc[Boolean mask, columns]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_130786</th>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>2.176788</td>\n",
       "      <td>1.998233</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_015665</th>\n",
       "      <td>8086.0</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>achalasia, adrenocortical insufficiency, alacr...</td>\n",
       "      <td>1.952652</td>\n",
       "      <td>3.753678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001087</th>\n",
       "      <td>14.0</td>\n",
       "      <td>AAMP</td>\n",
       "      <td>angio-associated, migratory cell protein</td>\n",
       "      <td>2.808783</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_012089</th>\n",
       "      <td>23456.0</td>\n",
       "      <td>ABCB10</td>\n",
       "      <td>ATP-binding cassette, sub-family B (MDR/TAP), ...</td>\n",
       "      <td>0.225564</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-4.433328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_014945</th>\n",
       "      <td>22885.0</td>\n",
       "      <td>ABLIM3</td>\n",
       "      <td>actin binding LIM protein family, member 3</td>\n",
       "      <td>2.205244</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_013375</th>\n",
       "      <td>29777.0</td>\n",
       "      <td>ABT1</td>\n",
       "      <td>activator of basal transcription 1</td>\n",
       "      <td>3.284837</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_004300</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.790544</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_177554</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.288103</td>\n",
       "      <td>1.219488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_007099</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.288103</td>\n",
       "      <td>1.219488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_203372</th>\n",
       "      <td>2181.0</td>\n",
       "      <td>ACSL3</td>\n",
       "      <td>acyl-CoA synthetase long-chain family member 3</td>\n",
       "      <td>0.254533</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-3.928758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_004457</th>\n",
       "      <td>2181.0</td>\n",
       "      <td>ACSL3</td>\n",
       "      <td>acyl-CoA synthetase long-chain family member 3</td>\n",
       "      <td>0.254533</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-3.928758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_005736</th>\n",
       "      <td>10121.0</td>\n",
       "      <td>ACTR1A</td>\n",
       "      <td>ARP1 actin-related protein 1 homolog A, centra...</td>\n",
       "      <td>2.088440</td>\n",
       "      <td>2.473819</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000022</th>\n",
       "      <td>100.0</td>\n",
       "      <td>ADA</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>3.166759</td>\n",
       "      <td>1.021824</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_003183</th>\n",
       "      <td>6868.0</td>\n",
       "      <td>ADAM17</td>\n",
       "      <td>ADAM metallopeptidase domain 17 (tumor necrosi...</td>\n",
       "      <td>0.416173</td>\n",
       "      <td>2.491259</td>\n",
       "      <td>-2.402845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_015999</th>\n",
       "      <td>51094.0</td>\n",
       "      <td>ADIPOR1</td>\n",
       "      <td>adiponectin receptor 1</td>\n",
       "      <td>1.894437</td>\n",
       "      <td>2.473819</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000644</th>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...</td>\n",
       "      <td>0.346660</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.884670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000643</th>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...</td>\n",
       "      <td>0.346660</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.884670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000646</th>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...</td>\n",
       "      <td>0.346660</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.884670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000028</th>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...</td>\n",
       "      <td>0.346660</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.884670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000642</th>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...</td>\n",
       "      <td>0.346660</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.884670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000645</th>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...</td>\n",
       "      <td>0.346660</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.884670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_003659</th>\n",
       "      <td>8540.0</td>\n",
       "      <td>AGPS</td>\n",
       "      <td>alkylglycerone phosphate synthase</td>\n",
       "      <td>2.216233</td>\n",
       "      <td>1.939343</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_006408</th>\n",
       "      <td>10551.0</td>\n",
       "      <td>AGR2</td>\n",
       "      <td>anterior gradient 2 homolog (Xenopus laevis)</td>\n",
       "      <td>0.372764</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-2.682662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_015446</th>\n",
       "      <td>25909.0</td>\n",
       "      <td>AHCTF1</td>\n",
       "      <td>AT hook containing transcription factor 1</td>\n",
       "      <td>0.240616</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-4.155992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001621</th>\n",
       "      <td>196.0</td>\n",
       "      <td>AHR</td>\n",
       "      <td>aryl hydrocarbon receptor</td>\n",
       "      <td>0.298388</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-3.351336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000476</th>\n",
       "      <td>203.0</td>\n",
       "      <td>AK1</td>\n",
       "      <td>adenylate kinase 1</td>\n",
       "      <td>2.667675</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_013411</th>\n",
       "      <td>204.0</td>\n",
       "      <td>AK2</td>\n",
       "      <td>adenylate kinase 2</td>\n",
       "      <td>2.246527</td>\n",
       "      <td>1.561308</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001625</th>\n",
       "      <td>204.0</td>\n",
       "      <td>AK2</td>\n",
       "      <td>adenylate kinase 2</td>\n",
       "      <td>3.030583</td>\n",
       "      <td>0.885581</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_016248</th>\n",
       "      <td>11215.0</td>\n",
       "      <td>AKAP11</td>\n",
       "      <td>A kinase (PRKA) anchor protein 11</td>\n",
       "      <td>0.446486</td>\n",
       "      <td>4.319905</td>\n",
       "      <td>-2.239711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_020299</th>\n",
       "      <td>57016.0</td>\n",
       "      <td>AKR1B10</td>\n",
       "      <td>aldo-keto reductase family 1, member B10 (aldo...</td>\n",
       "      <td>3.429254</td>\n",
       "      <td>4.319905</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_007233</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.695370</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_203353</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.521016</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_495965</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.227140</td>\n",
       "      <td>1.300108</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_376247</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.330743</td>\n",
       "      <td>3.950857</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_499012</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.963309</td>\n",
       "      <td>2.601228</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_531426</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.432487</td>\n",
       "      <td>3.235147</td>\n",
       "      <td>-2.312206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_497679</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.888995</td>\n",
       "      <td>3.753678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_379736</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.999065</td>\n",
       "      <td>3.797649</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_178225</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.754399</td>\n",
       "      <td>3.797649</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001002256</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.839806</td>\n",
       "      <td>3.797649</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_370878</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.546714</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>-1.829111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_374765</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.371331</td>\n",
       "      <td>1.998233</td>\n",
       "      <td>-2.693017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_004652</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.441336</td>\n",
       "      <td>3.753678</td>\n",
       "      <td>-2.265846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_021906</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.441336</td>\n",
       "      <td>3.753678</td>\n",
       "      <td>-2.265846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_145323</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.266313</td>\n",
       "      <td>2.016183</td>\n",
       "      <td>-3.754986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_016496</th>\n",
       "      <td>51257.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>membrane-associated ring finger (C3HC4) 2</td>\n",
       "      <td>2.220592</td>\n",
       "      <td>4.319905</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_211529</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.801124</td>\n",
       "      <td>4.319905</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_372143</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.019991</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_379250</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.057951</td>\n",
       "      <td>2.103816</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-HUMRGE/M10098_5_at</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.350925</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-2.849608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_374065</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.486665</td>\n",
       "      <td>2.491259</td>\n",
       "      <td>-2.054800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_379694</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.999065</td>\n",
       "      <td>3.797649</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_371848</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.970303</td>\n",
       "      <td>2.601228</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XR_000223</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.038178</td>\n",
       "      <td>4.934189</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_499128</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.022438</td>\n",
       "      <td>1.219488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_372719</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.349097</td>\n",
       "      <td>2.016183</td>\n",
       "      <td>-2.864532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_016572</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.102908</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_379274</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.172858</td>\n",
       "      <td>1.939343</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_114303</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.446474</td>\n",
       "      <td>2.590323</td>\n",
       "      <td>-2.239770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_496301</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.374709</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.668740</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1597 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Geneid Gene symbol  \\\n",
       "NM_130786                    1.0        A1BG   \n",
       "NM_015665                 8086.0        AAAS   \n",
       "NM_001087                   14.0        AAMP   \n",
       "NM_012089                23456.0      ABCB10   \n",
       "NM_014945                22885.0      ABLIM3   \n",
       "NM_013375                29777.0        ABT1   \n",
       "NM_004300                   52.0        ACP1   \n",
       "NM_177554                   52.0        ACP1   \n",
       "NM_007099                   52.0        ACP1   \n",
       "NM_203372                 2181.0       ACSL3   \n",
       "NM_004457                 2181.0       ACSL3   \n",
       "NM_005736                10121.0      ACTR1A   \n",
       "NM_000022                  100.0         ADA   \n",
       "NM_003183                 6868.0      ADAM17   \n",
       "NM_015999                51094.0     ADIPOR1   \n",
       "NM_000644                  178.0         AGL   \n",
       "NM_000643                  178.0         AGL   \n",
       "NM_000646                  178.0         AGL   \n",
       "NM_000028                  178.0         AGL   \n",
       "NM_000642                  178.0         AGL   \n",
       "NM_000645                  178.0         AGL   \n",
       "NM_003659                 8540.0        AGPS   \n",
       "NM_006408                10551.0        AGR2   \n",
       "NM_015446                25909.0      AHCTF1   \n",
       "NM_001621                  196.0         AHR   \n",
       "NM_000476                  203.0         AK1   \n",
       "NM_013411                  204.0         AK2   \n",
       "NM_001625                  204.0         AK2   \n",
       "NM_016248                11215.0      AKAP11   \n",
       "NM_020299                57016.0     AKR1B10   \n",
       "...                          ...         ...   \n",
       "NM_007233                    nan         NaN   \n",
       "NM_203353                    nan         NaN   \n",
       "XM_495965                    nan         NaN   \n",
       "XM_376247                    nan         NaN   \n",
       "XM_499012                    nan         NaN   \n",
       "XM_531426                    nan         NaN   \n",
       "XM_497679                    nan         NaN   \n",
       "XM_379736                    nan         NaN   \n",
       "NM_178225                    nan         NaN   \n",
       "NM_001002256                 nan         NaN   \n",
       "XM_370878                    nan         NaN   \n",
       "XM_374765                    nan         NaN   \n",
       "NM_004652                    nan         NaN   \n",
       "NM_021906                    nan         NaN   \n",
       "NM_145323                    nan         NaN   \n",
       "NM_016496                51257.0         NaN   \n",
       "XM_211529                    nan         NaN   \n",
       "XM_372143                    nan         NaN   \n",
       "XM_379250                    nan         NaN   \n",
       "AFFX-HUMRGE/M10098_5_at      nan         NaN   \n",
       "XM_374065                    nan         NaN   \n",
       "XM_379694                    nan         NaN   \n",
       "XM_371848                    nan         NaN   \n",
       "XR_000223                    nan         NaN   \n",
       "XM_499128                    nan         NaN   \n",
       "XM_372719                    nan         NaN   \n",
       "NM_016572                    nan         NaN   \n",
       "XM_379274                    nan         NaN   \n",
       "XM_114303                    nan         NaN   \n",
       "XM_496301                    nan         NaN   \n",
       "\n",
       "                                                                 Gene name  \\\n",
       "NM_130786                                           alpha-1-B glycoprotein   \n",
       "NM_015665                achalasia, adrenocortical insufficiency, alacr...   \n",
       "NM_001087                         angio-associated, migratory cell protein   \n",
       "NM_012089                ATP-binding cassette, sub-family B (MDR/TAP), ...   \n",
       "NM_014945                       actin binding LIM protein family, member 3   \n",
       "NM_013375                               activator of basal transcription 1   \n",
       "NM_004300                                      acid phosphatase 1, soluble   \n",
       "NM_177554                                      acid phosphatase 1, soluble   \n",
       "NM_007099                                      acid phosphatase 1, soluble   \n",
       "NM_203372                   acyl-CoA synthetase long-chain family member 3   \n",
       "NM_004457                   acyl-CoA synthetase long-chain family member 3   \n",
       "NM_005736                ARP1 actin-related protein 1 homolog A, centra...   \n",
       "NM_000022                                              adenosine deaminase   \n",
       "NM_003183                ADAM metallopeptidase domain 17 (tumor necrosi...   \n",
       "NM_015999                                           adiponectin receptor 1   \n",
       "NM_000644                amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...   \n",
       "NM_000643                amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...   \n",
       "NM_000646                amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...   \n",
       "NM_000028                amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...   \n",
       "NM_000642                amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...   \n",
       "NM_000645                amylo-1, 6-glucosidase, 4-alpha-glucanotransfe...   \n",
       "NM_003659                                alkylglycerone phosphate synthase   \n",
       "NM_006408                     anterior gradient 2 homolog (Xenopus laevis)   \n",
       "NM_015446                        AT hook containing transcription factor 1   \n",
       "NM_001621                                        aryl hydrocarbon receptor   \n",
       "NM_000476                                               adenylate kinase 1   \n",
       "NM_013411                                               adenylate kinase 2   \n",
       "NM_001625                                               adenylate kinase 2   \n",
       "NM_016248                                A kinase (PRKA) anchor protein 11   \n",
       "NM_020299                aldo-keto reductase family 1, member B10 (aldo...   \n",
       "...                                                                    ...   \n",
       "NM_007233                                                              NaN   \n",
       "NM_203353                                                              NaN   \n",
       "XM_495965                                                              NaN   \n",
       "XM_376247                                                              NaN   \n",
       "XM_499012                                                              NaN   \n",
       "XM_531426                                                              NaN   \n",
       "XM_497679                                                              NaN   \n",
       "XM_379736                                                              NaN   \n",
       "NM_178225                                                              NaN   \n",
       "NM_001002256                                                           NaN   \n",
       "XM_370878                                                              NaN   \n",
       "XM_374765                                                              NaN   \n",
       "NM_004652                                                              NaN   \n",
       "NM_021906                                                              NaN   \n",
       "NM_145323                                                              NaN   \n",
       "NM_016496                        membrane-associated ring finger (C3HC4) 2   \n",
       "XM_211529                                                              NaN   \n",
       "XM_372143                                                              NaN   \n",
       "XM_379250                                                              NaN   \n",
       "AFFX-HUMRGE/M10098_5_at                                                NaN   \n",
       "XM_374065                                                              NaN   \n",
       "XM_379694                                                              NaN   \n",
       "XM_371848                                                              NaN   \n",
       "XR_000223                                                              NaN   \n",
       "XM_499128                                                              NaN   \n",
       "XM_372719                                                              NaN   \n",
       "NM_016572                                                              NaN   \n",
       "XM_379274                                                              NaN   \n",
       "XM_114303                                                              NaN   \n",
       "XM_496301                                                              NaN   \n",
       "\n",
       "                         Fold Change  q-value (%)  Negative Fold Change  \n",
       "NM_130786                   2.176788     1.998233                   NaN  \n",
       "NM_015665                   1.952652     3.753678                   NaN  \n",
       "NM_001087                   2.808783     0.637618                   NaN  \n",
       "NM_012089                   0.225564     1.707905             -4.433328  \n",
       "NM_014945                   2.205244     4.611046                   NaN  \n",
       "NM_013375                   3.284837     0.637618                   NaN  \n",
       "NM_004300                   2.790544     0.637618                   NaN  \n",
       "NM_177554                   2.288103     1.219488                   NaN  \n",
       "NM_007099                   2.288103     1.219488                   NaN  \n",
       "NM_203372                   0.254533     1.707905             -3.928758  \n",
       "NM_004457                   0.254533     1.707905             -3.928758  \n",
       "NM_005736                   2.088440     2.473819                   NaN  \n",
       "NM_000022                   3.166759     1.021824                   NaN  \n",
       "NM_003183                   0.416173     2.491259             -2.402845  \n",
       "NM_015999                   1.894437     2.473819                   NaN  \n",
       "NM_000644                   0.346660     3.399167             -2.884670  \n",
       "NM_000643                   0.346660     3.399167             -2.884670  \n",
       "NM_000646                   0.346660     3.399167             -2.884670  \n",
       "NM_000028                   0.346660     3.399167             -2.884670  \n",
       "NM_000642                   0.346660     3.399167             -2.884670  \n",
       "NM_000645                   0.346660     3.399167             -2.884670  \n",
       "NM_003659                   2.216233     1.939343                   NaN  \n",
       "NM_006408                   0.372764     1.707905             -2.682662  \n",
       "NM_015446                   0.240616     1.707905             -4.155992  \n",
       "NM_001621                   0.298388     1.707905             -3.351336  \n",
       "NM_000476                   2.667675     0.637618                   NaN  \n",
       "NM_013411                   2.246527     1.561308                   NaN  \n",
       "NM_001625                   3.030583     0.885581                   NaN  \n",
       "NM_016248                   0.446486     4.319905             -2.239711  \n",
       "NM_020299                   3.429254     4.319905                   NaN  \n",
       "...                              ...          ...                   ...  \n",
       "NM_007233                   2.695370     0.637618                   NaN  \n",
       "NM_203353                   2.521016     0.637618                   NaN  \n",
       "XM_495965                   2.227140     1.300108                   NaN  \n",
       "XM_376247                   2.330743     3.950857                   NaN  \n",
       "XM_499012                   1.963309     2.601228                   NaN  \n",
       "XM_531426                   0.432487     3.235147             -2.312206  \n",
       "XM_497679                   1.888995     3.753678                   NaN  \n",
       "XM_379736                   1.999065     3.797649                   NaN  \n",
       "NM_178225                   1.754399     3.797649                   NaN  \n",
       "NM_001002256                1.839806     3.797649                   NaN  \n",
       "XM_370878                   0.546714     4.611046             -1.829111  \n",
       "XM_374765                   0.371331     1.998233             -2.693017  \n",
       "NM_004652                   0.441336     3.753678             -2.265846  \n",
       "NM_021906                   0.441336     3.753678             -2.265846  \n",
       "NM_145323                   0.266313     2.016183             -3.754986  \n",
       "NM_016496                   2.220592     4.319905                   NaN  \n",
       "XM_211529                   1.801124     4.319905                   NaN  \n",
       "XM_372143                   3.019991     0.637618                   NaN  \n",
       "XM_379250                   3.057951     2.103816                   NaN  \n",
       "AFFX-HUMRGE/M10098_5_at     0.350925     1.707905             -2.849608  \n",
       "XM_374065                   0.486665     2.491259             -2.054800  \n",
       "XM_379694                   1.999065     3.797649                   NaN  \n",
       "XM_371848                   1.970303     2.601228                   NaN  \n",
       "XR_000223                   2.038178     4.934189                   NaN  \n",
       "XM_499128                   2.022438     1.219488                   NaN  \n",
       "XM_372719                   0.349097     2.016183             -2.864532  \n",
       "NM_016572                   2.102908     4.611046                   NaN  \n",
       "XM_379274                   2.172858     1.939343                   NaN  \n",
       "XM_114303                   0.446474     2.590323             -2.239770  \n",
       "XM_496301                   0.374709     3.399167             -2.668740  \n",
       "\n",
       "[1597 rows x 6 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strict_frame = genelist.loc[mask,:]\n",
    "strict_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1597, 6)"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strict_frame.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just for fun I want to pull out the rows of the dataframe where the gene symbol is something specific like my labmates' 3 favorite genes.  You know, when your lab members ask \"What is the fold-change for this gene?\" and they for whatever reason don't know how to use the find fucntion in excel.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NM_130786    False\n",
       "NM_015665    False\n",
       "NM_001087    False\n",
       "NM_012089    False\n",
       "NM_014945    False\n",
       "Name: Gene symbol, dtype: bool"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sublist = ['ACP1', 'SOMEGENE', 'ADA']\n",
    "strict_frame['Gene symbol'].isin(sublist).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_004300</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.790544</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_177554</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.288103</td>\n",
       "      <td>1.219488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_007099</th>\n",
       "      <td>52.0</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>acid phosphatase 1, soluble</td>\n",
       "      <td>2.288103</td>\n",
       "      <td>1.219488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000022</th>\n",
       "      <td>100.0</td>\n",
       "      <td>ADA</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>3.166759</td>\n",
       "      <td>1.021824</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Geneid Gene symbol                    Gene name  Fold Change  \\\n",
       "NM_004300   52.0        ACP1  acid phosphatase 1, soluble     2.790544   \n",
       "NM_177554   52.0        ACP1  acid phosphatase 1, soluble     2.288103   \n",
       "NM_007099   52.0        ACP1  acid phosphatase 1, soluble     2.288103   \n",
       "NM_000022  100.0         ADA          adenosine deaminase     3.166759   \n",
       "\n",
       "           q-value (%)  Negative Fold Change  \n",
       "NM_004300     0.637618                   NaN  \n",
       "NM_177554     1.219488                   NaN  \n",
       "NM_007099     1.219488                   NaN  \n",
       "NM_000022     1.021824                   NaN  "
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#save the mask to a variable and use it to index your matrix\n",
    "newmask = strict_frame['Gene symbol'].isin(sublist)\n",
    "strict_frame.loc[newmask,:]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tools to filter our NaN's"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Things that aren't useful: Gene Symbols that are NaN.  So another way to make our dataset smaller is to filter out those NaNs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XM_372719</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.349097</td>\n",
       "      <td>2.016183</td>\n",
       "      <td>-2.864532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_016572</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.102908</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_379274</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.172858</td>\n",
       "      <td>1.939343</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_114303</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.446474</td>\n",
       "      <td>2.590323</td>\n",
       "      <td>-2.239770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_496301</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.374709</td>\n",
       "      <td>3.399167</td>\n",
       "      <td>-2.668740</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Geneid Gene symbol Gene name  Fold Change  q-value (%)  \\\n",
       "XM_372719    nan         NaN       NaN     0.349097     2.016183   \n",
       "NM_016572    nan         NaN       NaN     2.102908     4.611046   \n",
       "XM_379274    nan         NaN       NaN     2.172858     1.939343   \n",
       "XM_114303    nan         NaN       NaN     0.446474     2.590323   \n",
       "XM_496301    nan         NaN       NaN     0.374709     3.399167   \n",
       "\n",
       "           Negative Fold Change  \n",
       "XM_372719             -2.864532  \n",
       "NM_016572                   NaN  \n",
       "XM_379274                   NaN  \n",
       "XM_114303             -2.239770  \n",
       "XM_496301             -2.668740  "
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strict_frame.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is actually a tool to identify NaNs.  We are going to use it to specifically identify the location of NaN's in a column.  However, it can take an entire dataframe as an arguement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NM_130786    False\n",
       "NM_015665    False\n",
       "NM_001087    False\n",
       "NM_012089    False\n",
       "NM_014945    False\n",
       "Name: Gene symbol, dtype: bool"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.isnull(strict_frame['Gene symbol']).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to return a dataframe where rows that have Gene symbol as NaN are no longer there.  We can use Boolean indexing to do this.  What happens when we run this code:\n",
    "```python\n",
    "strict_frame.loc [pd.isnull(strict_frame['Gene symbol']),:]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_144495</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.383440</td>\n",
       "      <td>2.972720</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_375449</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.963692</td>\n",
       "      <td>3.672424</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XR_000132</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.682663</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_183063</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.127440</td>\n",
       "      <td>1.939343</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XM_375698</th>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.497546</td>\n",
       "      <td>2.601228</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Geneid Gene symbol Gene name  Fold Change  q-value (%)  \\\n",
       "NM_144495    nan         NaN       NaN     2.383440     2.972720   \n",
       "XM_375449    nan         NaN       NaN     1.963692     3.672424   \n",
       "XR_000132    nan         NaN       NaN     3.682663     0.637618   \n",
       "NM_183063    nan         NaN       NaN     2.127440     1.939343   \n",
       "XM_375698    nan         NaN       NaN     2.497546     2.601228   \n",
       "\n",
       "           Negative Fold Change  \n",
       "NM_144495                   NaN  \n",
       "XM_375449                   NaN  \n",
       "XR_000132                   NaN  \n",
       "NM_183063                   NaN  \n",
       "XM_375698                   NaN  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strict_frame.loc [pd.isnull(strict_frame['Gene symbol']),:].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You should notice that when you index with the Boolean mask, you return all of the rows where Gene Symbol has a NaN.  Which is actually the exact opposite of what we want to do, however, there's a nifty way of taking the inverse. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_130786</th>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>2.176788</td>\n",
       "      <td>1.998233</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_015665</th>\n",
       "      <td>8086.0</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>achalasia, adrenocortical insufficiency, alacr...</td>\n",
       "      <td>1.952652</td>\n",
       "      <td>3.753678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001087</th>\n",
       "      <td>14.0</td>\n",
       "      <td>AAMP</td>\n",
       "      <td>angio-associated, migratory cell protein</td>\n",
       "      <td>2.808783</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_012089</th>\n",
       "      <td>23456.0</td>\n",
       "      <td>ABCB10</td>\n",
       "      <td>ATP-binding cassette, sub-family B (MDR/TAP), ...</td>\n",
       "      <td>0.225564</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-4.433328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_014945</th>\n",
       "      <td>22885.0</td>\n",
       "      <td>ABLIM3</td>\n",
       "      <td>actin binding LIM protein family, member 3</td>\n",
       "      <td>2.205244</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Geneid Gene symbol  \\\n",
       "NM_130786      1.0        A1BG   \n",
       "NM_015665   8086.0        AAAS   \n",
       "NM_001087     14.0        AAMP   \n",
       "NM_012089  23456.0      ABCB10   \n",
       "NM_014945  22885.0      ABLIM3   \n",
       "\n",
       "                                                   Gene name  Fold Change  \\\n",
       "NM_130786                             alpha-1-B glycoprotein     2.176788   \n",
       "NM_015665  achalasia, adrenocortical insufficiency, alacr...     1.952652   \n",
       "NM_001087           angio-associated, migratory cell protein     2.808783   \n",
       "NM_012089  ATP-binding cassette, sub-family B (MDR/TAP), ...     0.225564   \n",
       "NM_014945         actin binding LIM protein family, member 3     2.205244   \n",
       "\n",
       "           q-value (%)  Negative Fold Change  \n",
       "NM_130786     1.998233                   NaN  \n",
       "NM_015665     3.753678                   NaN  \n",
       "NM_001087     0.637618                   NaN  \n",
       "NM_012089     1.707905             -4.433328  \n",
       "NM_014945     4.611046                   NaN  "
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stricter = strict_frame.loc[~pd.isnull(strict_frame['Gene symbol']),:]\n",
    "stricter.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1523, 6)"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stricter.shape\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering by Fold-change - Using And/Or Conditional Statement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not much reduction.  Let's try one more thing.  Let's figure out which genes have a 2 fold change in up or down direction.  Looking at our data, we notice that there are only positive numbers in the fold-change column.  If we look at our distribution plot for fold-change, we notice what about the distribution?:\n",
    "1. our positive fold-changes are above:\n",
    "2. our negative fold-changes are below:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a mask for upregulated genes and a separate mask for downregulated genes.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Geneid', 'Gene symbol', 'Gene name', 'Fold Change', 'q-value (%)',\n",
       "       'Negative Fold Change'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strict_frame.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "up = strict_frame['Fold Change'] > 2\n",
    "down = strict_frame['Fold Change']<0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Fold Change</th>\n",
       "      <th>q-value (%)</th>\n",
       "      <th>Negative Fold Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_130786</th>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>2.176788</td>\n",
       "      <td>1.998233</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_001087</th>\n",
       "      <td>14.0</td>\n",
       "      <td>AAMP</td>\n",
       "      <td>angio-associated, migratory cell protein</td>\n",
       "      <td>2.808783</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_012089</th>\n",
       "      <td>23456.0</td>\n",
       "      <td>ABCB10</td>\n",
       "      <td>ATP-binding cassette, sub-family B (MDR/TAP), ...</td>\n",
       "      <td>0.225564</td>\n",
       "      <td>1.707905</td>\n",
       "      <td>-4.433328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_014945</th>\n",
       "      <td>22885.0</td>\n",
       "      <td>ABLIM3</td>\n",
       "      <td>actin binding LIM protein family, member 3</td>\n",
       "      <td>2.205244</td>\n",
       "      <td>4.611046</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_013375</th>\n",
       "      <td>29777.0</td>\n",
       "      <td>ABT1</td>\n",
       "      <td>activator of basal transcription 1</td>\n",
       "      <td>3.284837</td>\n",
       "      <td>0.637618</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Geneid Gene symbol  \\\n",
       "NM_130786      1.0        A1BG   \n",
       "NM_001087     14.0        AAMP   \n",
       "NM_012089  23456.0      ABCB10   \n",
       "NM_014945  22885.0      ABLIM3   \n",
       "NM_013375  29777.0        ABT1   \n",
       "\n",
       "                                                   Gene name  Fold Change  \\\n",
       "NM_130786                             alpha-1-B glycoprotein     2.176788   \n",
       "NM_001087           angio-associated, migratory cell protein     2.808783   \n",
       "NM_012089  ATP-binding cassette, sub-family B (MDR/TAP), ...     0.225564   \n",
       "NM_014945         actin binding LIM protein family, member 3     2.205244   \n",
       "NM_013375                 activator of basal transcription 1     3.284837   \n",
       "\n",
       "           q-value (%)  Negative Fold Change  \n",
       "NM_130786     1.998233                   NaN  \n",
       "NM_001087     0.637618                   NaN  \n",
       "NM_012089     1.707905             -4.433328  \n",
       "NM_014945     4.611046                   NaN  \n",
       "NM_013375     0.637618                   NaN  "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#or statement\n",
    "strict_frame.loc[(up) | (down),:].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How can we create a mask that filters for q-value and fold change? Create one more mask that selects for rows that are below 5% - see above and use the genelist dataframe object to filter for q-value and fold change in one line of code.  This will require that you google how to filter rows with AND condition in pandas. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "qvalue_mask = genelist['q-value (%)']<5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "final_frame = strict_frame.loc[(up) | (down) & (genelist['q-value (%)']<5),:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Did this reduce the number of genes? How would we check?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1348, 6)"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_frame.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write your filtered data to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "strict_frame.to_csv('filtered_data.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For next time: We will be figuring out what to do with all of those duplicate values for gene symbols, and we will combine this current dataframe with perhaps a dictionary or other dataframe and plotting that data."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
